{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1beab998-5752-460c-88cc-811abeffc8e2",
   "metadata": {},
   "source": [
    "# Data Preprocessing\n",
    "\n",
    "## Overview\n",
    "\n",
    "The raw data from Eraslan et al. is a tab separated table containing gene names, related Ensembl IDs and measured or calculated values for mRNA abundance, protein abundance and protein-to-mRNA ratio.\n",
    "\n",
    "In the first Jupyter cells the data is roughly explored. After that the relevant values for the up coming analysis are extracted.\n",
    "\n",
    "Some problems arose:\n",
    "- Not all transcript IDs seem to be the current canonical form of transcript for a particular gene. In fact the given IDs point to transcripts that do not translate (in most cases). The gene needs to be identified in that case and the up to date transcript ID resolved.\n",
    "- BUT there are still some transcript IDs left that are annotate with \"nonsense mediated decay\". These will be thrown out as they do not successfully translate. \n",
    "- The Genecode data set contains duplicate files for some of the given transcript IDs. They can be easily filtered using a regex. The duplicates are the exact same files with different names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0fb525e9-e717-4129-99a3-89e5873b20e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# library dependencies\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e3b47d-618e-4b6c-9d5a-7ef18c8bbac4",
   "metadata": {},
   "source": [
    "## Reading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8e5f0e17-2ae6-42ec-b039-eeab545d756e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw data file and path\n",
    "datafile = '../data/Eraslan-EV3.tsv'\n",
    "\n",
    "# sanity check if the file exists\n",
    "if not Path(datafile).is_file():\n",
    "    print('Data file not found!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0edbe60a-daa9-419a-8635-184aeb0dda89",
   "metadata": {},
   "source": [
    "## Exploring the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1fc2f0e8-62f4-4b85-912e-50b04cc3d182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>EnsemblProteinID</th>\n",
       "      <th>Adrenal_mRNA</th>\n",
       "      <th>Appendices_mRNA</th>\n",
       "      <th>Brain_mRNA</th>\n",
       "      <th>Colon_mRNA</th>\n",
       "      <th>Duodenum_mRNA</th>\n",
       "      <th>Endometrium_mRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>ENSP00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.073</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.718</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.313</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>ENSP00000363105</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.971</td>\n",
       "      <td>2.324</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>5.147</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.202</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.8143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>ENSP00000323929</td>\n",
       "      <td>3.154</td>\n",
       "      <td>3.021</td>\n",
       "      <td>2.824</td>\n",
       "      <td>3.321</td>\n",
       "      <td>3.006</td>\n",
       "      <td>3.344</td>\n",
       "      <td>...</td>\n",
       "      <td>6.081</td>\n",
       "      <td>5.726</td>\n",
       "      <td>5.699</td>\n",
       "      <td>4.997</td>\n",
       "      <td>5.136</td>\n",
       "      <td>6.5349</td>\n",
       "      <td>5.820</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>ENSP00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.355</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>ENSP00000384794</td>\n",
       "      <td>1.625</td>\n",
       "      <td>1.567</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>4.731</td>\n",
       "      <td>4.508</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0613</td>\n",
       "      <td>4.832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>ENSP00000364023</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11571</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>ENSP00000294353</td>\n",
       "      <td>1.930</td>\n",
       "      <td>1.589</td>\n",
       "      <td>1.995</td>\n",
       "      <td>1.627</td>\n",
       "      <td>1.531</td>\n",
       "      <td>2.082</td>\n",
       "      <td>...</td>\n",
       "      <td>4.962</td>\n",
       "      <td>4.987</td>\n",
       "      <td>5.076</td>\n",
       "      <td>4.827</td>\n",
       "      <td>4.255</td>\n",
       "      <td>4.0412</td>\n",
       "      <td>5.389</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11572</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>ENSP00000324422</td>\n",
       "      <td>2.414</td>\n",
       "      <td>2.978</td>\n",
       "      <td>2.349</td>\n",
       "      <td>2.257</td>\n",
       "      <td>2.572</td>\n",
       "      <td>3.175</td>\n",
       "      <td>...</td>\n",
       "      <td>6.268</td>\n",
       "      <td>5.564</td>\n",
       "      <td>5.708</td>\n",
       "      <td>6.284</td>\n",
       "      <td>6.159</td>\n",
       "      <td>5.8846</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>ENSP00000371051</td>\n",
       "      <td>1.851</td>\n",
       "      <td>1.904</td>\n",
       "      <td>1.866</td>\n",
       "      <td>2.140</td>\n",
       "      <td>2.175</td>\n",
       "      <td>1.689</td>\n",
       "      <td>...</td>\n",
       "      <td>5.540</td>\n",
       "      <td>5.181</td>\n",
       "      <td>5.303</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.110</td>\n",
       "      <td>5.0834</td>\n",
       "      <td>5.047</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>ENSP00000359837</td>\n",
       "      <td>2.065</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.940</td>\n",
       "      <td>1.908</td>\n",
       "      <td>2.079</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.458</td>\n",
       "      <td>3.380</td>\n",
       "      <td>3.010</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.4766</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11575 rows × 91 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneName    EnsemblGeneID EnsemblTranscriptID EnsemblProteinID  \\\n",
       "0         A1BG  ENSG00000121410     ENST00000263100  ENSP00000263100   \n",
       "1         A1CF  ENSG00000148584     ENST00000373993  ENSP00000363105   \n",
       "2          A2M  ENSG00000175899     ENST00000318602  ENSP00000323929   \n",
       "3        A2ML1  ENSG00000166535     ENST00000299698  ENSP00000299698   \n",
       "4       A4GALT  ENSG00000128274     ENST00000401850  ENSP00000384794   \n",
       "...        ...              ...                 ...              ...   \n",
       "11570     ZXDB  ENSG00000198455     ENST00000374888  ENSP00000364023   \n",
       "11571   ZYG11B  ENSG00000162378     ENST00000294353  ENSP00000294353   \n",
       "11572      ZYX  ENSG00000159840     ENST00000322764  ENSP00000324422   \n",
       "11573    ZZEF1  ENSG00000074755     ENST00000381638  ENSP00000371051   \n",
       "11574     ZZZ3  ENSG00000036549     ENST00000370801  ENSP00000359837   \n",
       "\n",
       "      Adrenal_mRNA Appendices_mRNA Brain_mRNA Colon_mRNA Duodenum_mRNA  \\\n",
       "0               NA           1.073         NA         NA            NA   \n",
       "1               NA              NA         NA      1.971         2.324   \n",
       "2            3.154           3.021      2.824      3.321         3.006   \n",
       "3               NA              NA      1.355         NA            NA   \n",
       "4            1.625           1.567         NA         NA            NA   \n",
       "...            ...             ...        ...        ...           ...   \n",
       "11570           NA              NA         NA         NA            NA   \n",
       "11571        1.930           1.589      1.995      1.627         1.531   \n",
       "11572        2.414           2.978      2.349      2.257         2.572   \n",
       "11573        1.851           1.904      1.866      2.140         2.175   \n",
       "11574        2.065              NA         NA      1.940         1.908   \n",
       "\n",
       "      Endometrium_mRNA  ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR  \\\n",
       "0                   NA  ...         NA             7.718                 NA   \n",
       "1                   NA  ...      5.147                NA              5.202   \n",
       "2                3.344  ...      6.081             5.726              5.699   \n",
       "3                   NA  ...         NA                NA                 NA   \n",
       "4                   NA  ...      4.731             4.508                 NA   \n",
       "...                ...  ...        ...               ...                ...   \n",
       "11570               NA  ...         NA                NA                 NA   \n",
       "11571            2.082  ...      4.962             4.987              5.076   \n",
       "11572            3.175  ...      6.268             5.564              5.708   \n",
       "11573            1.689  ...      5.540             5.181              5.303   \n",
       "11574            2.079  ...         NA             3.458              3.380   \n",
       "\n",
       "      Smoothmuscle_PTR Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR  \\\n",
       "0                   NA      7.313          NA         NA          NA   \n",
       "1                   NA         NA      5.8143         NA          NA   \n",
       "2                4.997      5.136      6.5349      5.820       6.060   \n",
       "3                   NA         NA          NA      2.350          NA   \n",
       "4                   NA         NA      4.0613      4.832          NA   \n",
       "...                ...        ...         ...        ...         ...   \n",
       "11570               NA         NA          NA         NA          NA   \n",
       "11571            4.827      4.255      4.0412      5.389       4.250   \n",
       "11572            6.284      6.159      5.8846      5.582       5.598   \n",
       "11573            5.038      5.110      5.0834      5.047       5.038   \n",
       "11574            3.010         NA      3.4766         NA          NA   \n",
       "\n",
       "      Tonsil_PTR Urinarybladder_PTR  \n",
       "0             NA                 NA  \n",
       "1             NA                 NA  \n",
       "2          5.675             5.8286  \n",
       "3          5.249                 NA  \n",
       "4             NA             4.2430  \n",
       "...          ...                ...  \n",
       "11570      4.681                 NA  \n",
       "11571      4.439             4.1460  \n",
       "11572      5.968             5.3358  \n",
       "11573      5.130             5.0619  \n",
       "11574      3.008             3.6041  \n",
       "\n",
       "[11575 rows x 91 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading the data into a dataframe and looking at the first entries\n",
    "df = pd.read_csv(datafile, sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10870254-b393-42b1-8a8d-7f38592b986b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 11575 entries, 0 to 11574\n",
      "Data columns (total 91 columns):\n",
      " #   Column                  Non-Null Count  Dtype \n",
      "---  ------                  --------------  ----- \n",
      " 0   GeneName                11575 non-null  object\n",
      " 1   EnsemblGeneID           11575 non-null  object\n",
      " 2   EnsemblTranscriptID     11575 non-null  object\n",
      " 3   EnsemblProteinID        11575 non-null  object\n",
      " 4   Adrenal_mRNA            11575 non-null  object\n",
      " 5   Appendices_mRNA         11575 non-null  object\n",
      " 6   Brain_mRNA              11575 non-null  object\n",
      " 7   Colon_mRNA              11575 non-null  object\n",
      " 8   Duodenum_mRNA           11575 non-null  object\n",
      " 9   Endometrium_mRNA        11575 non-null  object\n",
      " 10  Esophagus_mRNA          11575 non-null  object\n",
      " 11  Fallopiantube_mRNA      11575 non-null  object\n",
      " 12  Fat_mRNA                11575 non-null  object\n",
      " 13  Gallbladder_mRNA        11575 non-null  object\n",
      " 14  Heart_mRNA              11575 non-null  object\n",
      " 15  Kidney_mRNA             11575 non-null  object\n",
      " 16  Liver_mRNA              11575 non-null  object\n",
      " 17  Lung_mRNA               11575 non-null  object\n",
      " 18  Lymphnode_mRNA          11575 non-null  object\n",
      " 19  Ovary_mRNA              11575 non-null  object\n",
      " 20  Pancreas_mRNA           11575 non-null  object\n",
      " 21  Placenta_mRNA           11575 non-null  object\n",
      " 22  Prostate_mRNA           11575 non-null  object\n",
      " 23  Rectum_mRNA             11575 non-null  object\n",
      " 24  Salivarygland_mRNA      11575 non-null  object\n",
      " 25  Smallintestine_mRNA     11575 non-null  object\n",
      " 26  Smoothmuscle_mRNA       11575 non-null  object\n",
      " 27  Spleen_mRNA             11575 non-null  object\n",
      " 28  Stomach_mRNA            11575 non-null  object\n",
      " 29  Testis_mRNA             11575 non-null  object\n",
      " 30  Thyroid_mRNA            11575 non-null  object\n",
      " 31  Tonsil_mRNA             11575 non-null  object\n",
      " 32  Urinarybladder_mRNA     11575 non-null  object\n",
      " 33  Adrenal_protein         11575 non-null  object\n",
      " 34  Appendices_protein      11575 non-null  object\n",
      " 35  Brain_protein           11575 non-null  object\n",
      " 36  Colon_protein           11575 non-null  object\n",
      " 37  Duodenum_protein        11575 non-null  object\n",
      " 38  Endometrium_protein     11575 non-null  object\n",
      " 39  Esophagus_protein       11575 non-null  object\n",
      " 40  Fallopiantube_protein   11575 non-null  object\n",
      " 41  Fat_protein             11575 non-null  object\n",
      " 42  Gallbladder_protein     11575 non-null  object\n",
      " 43  Heart_protein           11575 non-null  object\n",
      " 44  Kidney_protein          11575 non-null  object\n",
      " 45  Liver_protein           11575 non-null  object\n",
      " 46  Lung_protein            11575 non-null  object\n",
      " 47  Lymphnode_protein       11575 non-null  object\n",
      " 48  Ovary_protein           11575 non-null  object\n",
      " 49  Pancreas_protein        11575 non-null  object\n",
      " 50  Placenta_protein        11575 non-null  object\n",
      " 51  Prostate_protein        11575 non-null  object\n",
      " 52  Rectum_protein          11575 non-null  object\n",
      " 53  Salivarygland_protein   11575 non-null  object\n",
      " 54  Smallintestine_protein  11575 non-null  object\n",
      " 55  Smoothmuscle_protein    11575 non-null  object\n",
      " 56  Spleen_protein          11575 non-null  object\n",
      " 57  Stomach_protein         11575 non-null  object\n",
      " 58  Testis_protein          11575 non-null  object\n",
      " 59  Thyroid_protein         11575 non-null  object\n",
      " 60  Tonsil_protein          11575 non-null  object\n",
      " 61  Urinarybladder_protein  11575 non-null  object\n",
      " 62  Adrenal_PTR             11575 non-null  object\n",
      " 63  Appendices_PTR          11575 non-null  object\n",
      " 64  Brain_PTR               11575 non-null  object\n",
      " 65  Colon_PTR               11575 non-null  object\n",
      " 66  Duodenum_PTR            11575 non-null  object\n",
      " 67  Endometrium_PTR         11575 non-null  object\n",
      " 68  Esophagus_PTR           11575 non-null  object\n",
      " 69  Fallopiantube_PTR       11575 non-null  object\n",
      " 70  Fat_PTR                 11575 non-null  object\n",
      " 71  Gallbladder_PTR         11575 non-null  object\n",
      " 72  Heart_PTR               11575 non-null  object\n",
      " 73  Kidney_PTR              11575 non-null  object\n",
      " 74  Liver_PTR               11575 non-null  object\n",
      " 75  Lung_PTR                11575 non-null  object\n",
      " 76  Lymphnode_PTR           11575 non-null  object\n",
      " 77  Ovary_PTR               11575 non-null  object\n",
      " 78  Pancreas_PTR            11575 non-null  object\n",
      " 79  Placenta_PTR            11575 non-null  object\n",
      " 80  Prostate_PTR            11575 non-null  object\n",
      " 81  Rectum_PTR              11575 non-null  object\n",
      " 82  Salivarygland_PTR       11575 non-null  object\n",
      " 83  Smallintestine_PTR      11575 non-null  object\n",
      " 84  Smoothmuscle_PTR        11575 non-null  object\n",
      " 85  Spleen_PTR              11575 non-null  object\n",
      " 86  Stomach_PTR             11575 non-null  object\n",
      " 87  Testis_PTR              11575 non-null  object\n",
      " 88  Thyroid_PTR             11575 non-null  object\n",
      " 89  Tonsil_PTR              11575 non-null  object\n",
      " 90  Urinarybladder_PTR      11575 non-null  object\n",
      "dtypes: object(91)\n",
      "memory usage: 8.0+ MB\n"
     ]
    }
   ],
   "source": [
    "# looking at the columns\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0ee1e23d-728a-4243-8050-f6b999fa2cbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>EnsemblProteinID</th>\n",
       "      <th>Adrenal_mRNA</th>\n",
       "      <th>Appendices_mRNA</th>\n",
       "      <th>Brain_mRNA</th>\n",
       "      <th>Colon_mRNA</th>\n",
       "      <th>Duodenum_mRNA</th>\n",
       "      <th>Endometrium_mRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>...</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>2132</td>\n",
       "      <td>2049</td>\n",
       "      <td>2020</td>\n",
       "      <td>2143</td>\n",
       "      <td>2183</td>\n",
       "      <td>2043</td>\n",
       "      <td>...</td>\n",
       "      <td>3175</td>\n",
       "      <td>3297</td>\n",
       "      <td>3117</td>\n",
       "      <td>3248</td>\n",
       "      <td>3109</td>\n",
       "      <td>7372</td>\n",
       "      <td>3299</td>\n",
       "      <td>3238</td>\n",
       "      <td>3102</td>\n",
       "      <td>6971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>ENSP00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3162</td>\n",
       "      <td>4011</td>\n",
       "      <td>3079</td>\n",
       "      <td>3396</td>\n",
       "      <td>2971</td>\n",
       "      <td>3395</td>\n",
       "      <td>...</td>\n",
       "      <td>3471</td>\n",
       "      <td>3643</td>\n",
       "      <td>2981</td>\n",
       "      <td>3632</td>\n",
       "      <td>4022</td>\n",
       "      <td>3124</td>\n",
       "      <td>2938</td>\n",
       "      <td>3962</td>\n",
       "      <td>3603</td>\n",
       "      <td>3697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 91 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneName    EnsemblGeneID EnsemblTranscriptID EnsemblProteinID  \\\n",
       "count     11575            11575               11575            11575   \n",
       "unique    11575            11575               11575            11575   \n",
       "top        A1BG  ENSG00000121410     ENST00000263100  ENSP00000263100   \n",
       "freq          1                1                   1                1   \n",
       "\n",
       "       Adrenal_mRNA Appendices_mRNA Brain_mRNA Colon_mRNA Duodenum_mRNA  \\\n",
       "count         11575           11575      11575      11575         11575   \n",
       "unique         2132            2049       2020       2143          2183   \n",
       "top              NA              NA         NA         NA            NA   \n",
       "freq           3162            4011       3079       3396          2971   \n",
       "\n",
       "       Endometrium_mRNA  ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR  \\\n",
       "count             11575  ...      11575             11575              11575   \n",
       "unique             2043  ...       3175              3297               3117   \n",
       "top                  NA  ...         NA                NA                 NA   \n",
       "freq               3395  ...       3471              3643               2981   \n",
       "\n",
       "       Smoothmuscle_PTR Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR  \\\n",
       "count             11575      11575       11575      11575       11575   \n",
       "unique             3248       3109        7372       3299        3238   \n",
       "top                  NA         NA          NA         NA          NA   \n",
       "freq               3632       4022        3124       2938        3962   \n",
       "\n",
       "       Tonsil_PTR Urinarybladder_PTR  \n",
       "count       11575              11575  \n",
       "unique       3102               6971  \n",
       "top            NA                 NA  \n",
       "freq         3603               3697  \n",
       "\n",
       "[4 rows x 91 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de10a03-6d3e-4348-a970-1394d58ef4c6",
   "metadata": {},
   "source": [
    "Between ~3000 and ~4000 values in each of the 11575 rows are NA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "049d92a4-c181-43b9-81db-e2fbd4a7a9cd",
   "metadata": {},
   "source": [
    "## Extracting the relevant columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a7b83b9-4921-422f-b99e-622e830e56d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.277</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.841</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.718</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.313</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.135</td>\n",
       "      <td>5.371</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>5.147</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.202</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.8143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>6.290</td>\n",
       "      <td>6.328</td>\n",
       "      <td>5.948</td>\n",
       "      <td>5.811</td>\n",
       "      <td>6.068</td>\n",
       "      <td>5.383</td>\n",
       "      <td>5.881</td>\n",
       "      <td>6.119</td>\n",
       "      <td>6.410</td>\n",
       "      <td>...</td>\n",
       "      <td>6.081</td>\n",
       "      <td>5.726</td>\n",
       "      <td>5.699</td>\n",
       "      <td>4.997</td>\n",
       "      <td>5.136</td>\n",
       "      <td>6.5349</td>\n",
       "      <td>5.820</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.995</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.129</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>3.843</td>\n",
       "      <td>4.601</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.013</td>\n",
       "      <td>3.683</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>4.731</td>\n",
       "      <td>4.508</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0613</td>\n",
       "      <td>4.832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11571</th>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>4.461</td>\n",
       "      <td>5.013</td>\n",
       "      <td>5.047</td>\n",
       "      <td>4.566</td>\n",
       "      <td>5.184</td>\n",
       "      <td>4.826</td>\n",
       "      <td>5.102</td>\n",
       "      <td>4.670</td>\n",
       "      <td>5.756</td>\n",
       "      <td>...</td>\n",
       "      <td>4.962</td>\n",
       "      <td>4.987</td>\n",
       "      <td>5.076</td>\n",
       "      <td>4.827</td>\n",
       "      <td>4.255</td>\n",
       "      <td>4.0412</td>\n",
       "      <td>5.389</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11572</th>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.524</td>\n",
       "      <td>5.478</td>\n",
       "      <td>5.915</td>\n",
       "      <td>5.811</td>\n",
       "      <td>5.817</td>\n",
       "      <td>5.943</td>\n",
       "      <td>5.509</td>\n",
       "      <td>4.931</td>\n",
       "      <td>...</td>\n",
       "      <td>6.268</td>\n",
       "      <td>5.564</td>\n",
       "      <td>5.708</td>\n",
       "      <td>6.284</td>\n",
       "      <td>6.159</td>\n",
       "      <td>5.8846</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>5.112</td>\n",
       "      <td>4.918</td>\n",
       "      <td>5.139</td>\n",
       "      <td>5.190</td>\n",
       "      <td>5.442</td>\n",
       "      <td>5.602</td>\n",
       "      <td>4.715</td>\n",
       "      <td>4.956</td>\n",
       "      <td>5.033</td>\n",
       "      <td>...</td>\n",
       "      <td>5.540</td>\n",
       "      <td>5.181</td>\n",
       "      <td>5.303</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.110</td>\n",
       "      <td>5.0834</td>\n",
       "      <td>5.047</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>3.165</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.539</td>\n",
       "      <td>3.629</td>\n",
       "      <td>3.616</td>\n",
       "      <td>4.203</td>\n",
       "      <td>3.400</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.458</td>\n",
       "      <td>3.380</td>\n",
       "      <td>3.010</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.4766</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11575 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EnsemblTranscriptID Adrenal_PTR Appendices_PTR Brain_PTR Colon_PTR  \\\n",
       "0         ENST00000263100          NA          8.277        NA        NA   \n",
       "1         ENST00000373993          NA             NA        NA     5.135   \n",
       "2         ENST00000318602       6.290          6.328     5.948     5.811   \n",
       "3         ENST00000299698          NA             NA     3.995        NA   \n",
       "4         ENST00000401850       3.843          4.601        NA        NA   \n",
       "...                   ...         ...            ...       ...       ...   \n",
       "11570     ENST00000374888          NA             NA        NA        NA   \n",
       "11571     ENST00000294353       4.461          5.013     5.047     4.566   \n",
       "11572     ENST00000322764       5.664          5.524     5.478     5.915   \n",
       "11573     ENST00000381638       5.112          4.918     5.139     5.190   \n",
       "11574     ENST00000370801       3.165             NA        NA     3.539   \n",
       "\n",
       "      Duodenum_PTR Endometrium_PTR Esophagus_PTR Fallopiantube_PTR Fat_PTR  \\\n",
       "0               NA              NA            NA             7.841      NA   \n",
       "1            5.371              NA            NA                NA      NA   \n",
       "2            6.068           5.383         5.881             6.119   6.410   \n",
       "3               NA              NA         4.129                NA      NA   \n",
       "4               NA              NA         4.013             3.683      NA   \n",
       "...            ...             ...           ...               ...     ...   \n",
       "11570           NA              NA            NA                NA      NA   \n",
       "11571        5.184           4.826         5.102             4.670   5.756   \n",
       "11572        5.811           5.817         5.943             5.509   4.931   \n",
       "11573        5.442           5.602         4.715             4.956   5.033   \n",
       "11574        3.629           3.616         4.203             3.400      NA   \n",
       "\n",
       "       ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR Smoothmuscle_PTR  \\\n",
       "0      ...         NA             7.718                 NA               NA   \n",
       "1      ...      5.147                NA              5.202               NA   \n",
       "2      ...      6.081             5.726              5.699            4.997   \n",
       "3      ...         NA                NA                 NA               NA   \n",
       "4      ...      4.731             4.508                 NA               NA   \n",
       "...    ...        ...               ...                ...              ...   \n",
       "11570  ...         NA                NA                 NA               NA   \n",
       "11571  ...      4.962             4.987              5.076            4.827   \n",
       "11572  ...      6.268             5.564              5.708            6.284   \n",
       "11573  ...      5.540             5.181              5.303            5.038   \n",
       "11574  ...         NA             3.458              3.380            3.010   \n",
       "\n",
       "      Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR Tonsil_PTR  \\\n",
       "0          7.313          NA         NA          NA         NA   \n",
       "1             NA      5.8143         NA          NA         NA   \n",
       "2          5.136      6.5349      5.820       6.060      5.675   \n",
       "3             NA          NA      2.350          NA      5.249   \n",
       "4             NA      4.0613      4.832          NA         NA   \n",
       "...          ...         ...        ...         ...        ...   \n",
       "11570         NA          NA         NA          NA      4.681   \n",
       "11571      4.255      4.0412      5.389       4.250      4.439   \n",
       "11572      6.159      5.8846      5.582       5.598      5.968   \n",
       "11573      5.110      5.0834      5.047       5.038      5.130   \n",
       "11574         NA      3.4766         NA          NA      3.008   \n",
       "\n",
       "      Urinarybladder_PTR  \n",
       "0                     NA  \n",
       "1                     NA  \n",
       "2                 5.8286  \n",
       "3                     NA  \n",
       "4                 4.2430  \n",
       "...                  ...  \n",
       "11570                 NA  \n",
       "11571             4.1460  \n",
       "11572             5.3358  \n",
       "11573             5.0619  \n",
       "11574             3.6041  \n",
       "\n",
       "[11575 rows x 30 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df[['EnsemblTranscriptID'] + [ col for col in df.columns if col.endswith('_PTR') ]].copy()\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76fe860-fd42-43ec-9818-0ffab352701e",
   "metadata": {},
   "source": [
    "Cross referencing the transcript IDs with BED and Fasta files from the Genecode data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5acf60f0-d6e7-4955-b214-23454e27a919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000435683 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263817 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370449 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000376887 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000260645 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622407 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000331789 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000366779 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355413 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000373176 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000313871 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000313871.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000313871.9:0-3199.fasta\n",
      "ENST00000564546 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000374855 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371108 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000319144 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330258 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620652 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355272 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622409 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000348624 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611099 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000431282 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000422706 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000540651 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378004 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620747 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611208 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000538189 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381317 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381317.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381317.9:0-2065.fasta\n",
      "ENST00000617719 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000537989 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370580 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000443071 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000335016 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371265 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000383778 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000286067 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622257 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000559313 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000542285 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000333059 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000535057 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000224181 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000382387 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309893 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000264122 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000270279 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000624406 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000442201 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392347 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618014 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000401743 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381192 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381192.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381192.10:0-1129.fasta\n",
      "ENST00000453495 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615636 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000260595 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619537 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620528 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000609024 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619275 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000248228 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620463 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545689 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265379 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000264193 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000215939 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000432318 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000432318.8.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000432318.8:0-1940.fasta\n",
      "ENST00000262825 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000315684 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372621 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611509 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409414 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545947 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000361951 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396817 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000539042 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000507955 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000397239 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409775 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000266991 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000334651 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000334651.11.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000334651.11:0-2579.fasta\n",
      "ENST00000398557 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000585328 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000342382 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000256935 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000339659 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618169 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000563459 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000334286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000298049 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000245323 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000269349 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621148 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615377 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613079 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000251101 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000360971 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000472938 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000389562 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621003 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613142 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612797 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000377828 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556177 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620555 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396472 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622754 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368695 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000540185 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000427718 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000369384 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613418 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000302053 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000443566 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000380874 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000422223 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000310160 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000505428 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621232 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000628124 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000568107 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622544 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000229416 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000322315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620807 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611961 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409774 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368232 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620890 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612794 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622181 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000251808 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000216951 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000326153 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000326153.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000326153.10:0-1875.fasta\n",
      "ENST00000380299 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000557538 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613394 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000256216 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372425 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263642 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558762 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000389983 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000412504 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381469 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381469.7.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381469.7:0-1476.fasta\n",
      "ENST00000401393 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000361976 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396435 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000379389 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000287497 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622834 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378565 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000543780 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368089 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000262888 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000406396 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355193 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265023 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614544 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612396 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000415258 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000592991 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541779 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000311322 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263816 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000500728 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000418342 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000395528 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000215739 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000348428 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000226578 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000493901 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000523944 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000629399 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000599213 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000434178 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000537344 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396617 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000591601 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000281317 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000233616 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000380172 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000216605 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000625889 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000359786 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265944 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541942 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368063 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000331837 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378062 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000411638 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000627621 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371527 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622383 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000436936 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000344089 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000277541 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620134 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000405342 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000376214 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381297 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381297.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381297.10:0-4267.fasta\n",
      "ENST00000379806 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000542749 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370184 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372389 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614426 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000404875 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000303531 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615510 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000357304 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000632805 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000606149 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620216 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619580 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000306726 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000403432 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000527788 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000329286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610732 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000343115 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000625761 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611665 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617334 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617441 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556710 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000318842 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000424325 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000464167 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368567 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534062 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000590320 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617721 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611159 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621965 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000377475 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541726 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611680 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611222 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615547 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616787 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612212 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367104 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616363 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330479 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265735 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000357436 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534261 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558405 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000298472 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000397260 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000273158 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381401 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381401.11.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381401.11:0-1451.fasta\n",
      "ENST00000427851 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392845 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000424970 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000539553 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000400286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614072 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000602728 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556626 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000342756 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000586243 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000267540 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378654 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000538427 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615887 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000432496 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616883 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000609199 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611869 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000536810 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000535286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613071 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615733 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558083 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392129 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617947 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000588847 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392451 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000406386 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000290846 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000454289 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622375 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368563 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000429002 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610664 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367924 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367274 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614096 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616789 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000286448 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000286448.12.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000286448.12:0-2594.fasta\n",
      "ENST00000261517 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000369130 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618527 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000434753 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000400602 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381223 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381223.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381223.9:0-4510.fasta\n",
      "ENST00000373452 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545588 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309776 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534834 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309495 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000543588 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "searched for 11575 and found 11281\n",
      "missing or otherwise off: 294\n"
     ]
    }
   ],
   "source": [
    "# raw data paths\n",
    "gencode_path = '../../GENCODE43/protein_coding/'\n",
    "bed = Path(gencode_path) / 'BED6__protein_coding_strict/'\n",
    "fa = Path(gencode_path) / 'FA_protein_coding_strict_mRNA/'\n",
    "\n",
    "# file names look like this\n",
    "# for the BED file : ENST00000370801.8.bed\n",
    "# for the Fasta file : ENST00000370801.8:0-6412.fasta\n",
    "# .8 denotes the current Ensemble version\n",
    "# :0-6412 is the nucleotide length\n",
    "\n",
    "count_all = 0\n",
    "count_found = 0\n",
    "\n",
    "# extend the dataframe\n",
    "df2['bed_files'] = 0\n",
    "df2['fa_files'] = 0\n",
    "df2['bed'] = ''\n",
    "df2['fa'] = ''\n",
    "\n",
    "# checking if all the transcript Fasta and BED files per transcript exist\n",
    "for tid in df2['EnsemblTranscriptID']:\n",
    "    # inclrease over all count\n",
    "    count_all += 1\n",
    "\n",
    "    # list and count files\n",
    "    bed_file_list = list(bed.glob(tid + '*.bed'))\n",
    "    bed_file_count = len(bed_file_list)\n",
    "    fa_file_list = list(fa.glob(tid + '*.fasta'))\n",
    "    fa_file_count = len(fa_file_list)\n",
    "\n",
    "    # update dataframe\n",
    "    df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed_files'] = bed_file_count\n",
    "    df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa_files'] = fa_file_count\n",
    "\n",
    "    # check BED and Fasta file count\n",
    "    if bed_file_count == 1 and fa_file_count == 1:\n",
    "        # exctly one BED and FA file\n",
    "        \n",
    "        # increase hit count\n",
    "        count_found += 1\n",
    "        \n",
    "        # update file name information\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed'] = str(bed_file_list[0])\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa'] = str(fa_file_list[0])\n",
    "    elif bed_file_count == 2 and fa_file_count == 2:\n",
    "        # special case where there are duplicate files\n",
    "        print(tid, 'more than one BED/Fasta file present. selecting')\n",
    "\n",
    "        # increase hit count\n",
    "        count_found += 1\n",
    "\n",
    "        # find correct BED file and update table\n",
    "        for f in bed_file_list:\n",
    "            temp_bed_file = str(f)\n",
    "            if re.search(r'.*ENST\\d+\\.\\d+.bed', temp_bed_file):\n",
    "                df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed'] = temp_bed_file\n",
    "                print('   ', temp_bed_file)\n",
    "\n",
    "        # find correct Fasta file and update table\n",
    "        for f in fa_file_list:\n",
    "            temp_fa_file = str(f)\n",
    "            if re.search(r'.*ENST\\d+\\.\\d+:\\d+-\\d+.fasta', temp_fa_file):\n",
    "                df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa'] = temp_fa_file\n",
    "                print('   ', temp_fa_file)\n",
    "\n",
    "        # update file count in table\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed_files'] = 1\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa_files'] = 1\n",
    "    else:\n",
    "        # everything else ends up here\n",
    "        print(tid, 'bed count:', bed_file_count, 'fa count:', fa_file_count, 'bed files:', bed_file_list, 'fa files:', fa_file_list)\n",
    "\n",
    "print('searched for', count_all, 'and found', count_found)\n",
    "print('missing or otherwise off:', count_all - count_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ba103ec0-a297-43fe-a45d-ba6d1b8b493e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "      <th>bed_files</th>\n",
       "      <th>fa_files</th>\n",
       "      <th>bed</th>\n",
       "      <th>fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [EnsemblTranscriptID, Adrenal_PTR, Appendices_PTR, Brain_PTR, Colon_PTR, Duodenum_PTR, Endometrium_PTR, Esophagus_PTR, Fallopiantube_PTR, Fat_PTR, Gallbladder_PTR, Heart_PTR, Kidney_PTR, Liver_PTR, Lung_PTR, Lymphnode_PTR, Ovary_PTR, Pancreas_PTR, Placenta_PTR, Prostate_PTR, Rectum_PTR, Salivarygland_PTR, Smallintestine_PTR, Smoothmuscle_PTR, Spleen_PTR, Stomach_PTR, Testis_PTR, Thyroid_PTR, Tonsil_PTR, Urinarybladder_PTR, bed_files, fa_files, bed, fa]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 34 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# entries with two transcript files per entry\n",
    "df2.loc[ df2['bed_files'] == 2 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53cd8bba-8966-4f95-98c4-b5942e0b07be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "294"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# missing files\n",
    "df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7cda9f82-6566-4955-b960-c675b07cbf62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28       ENST00000435683\n",
       "33       ENST00000263817\n",
       "39       ENST00000370449\n",
       "41       ENST00000376887\n",
       "54       ENST00000260645\n",
       "              ...       \n",
       "11323    ENST00000545588\n",
       "11411    ENST00000309776\n",
       "11441    ENST00000534834\n",
       "11513    ENST00000309495\n",
       "11559    ENST00000543588\n",
       "Name: EnsemblTranscriptID, Length: 294, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ae332d-3367-460b-b5f5-08ac4fb3712d",
   "metadata": {},
   "outputs": [],
   "source": [
    "URL = 'https://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=ENST00000263817'\n",
    "r = requests.get(URL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d81e4fcd-5d16-41cd-9a3f-6c450e477e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "soup = BeautifulSoup(r.content, 'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b2941d-3f23-4c3f-a6e7-21ebc1f0e25f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if re.search(r'This transcript is not in the current gene set', soup.get_text()):\n",
    "    print('deprecated')\n",
    "    href = soup.td.next_sibling.a.attrs['href']\n",
    "    gene = re.sub(r'.*(ENSG0\\d+)', r'\\1', href)\n",
    "    print(href)\n",
    "    print(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d1247d-c346-4772-8a04-576115c148a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "URL = 'https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=ENSG00000073734'\n",
    "r = requests.get(URL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c92124-7865-4396-af80-5d3d98cc8fd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "soup = BeautifulSoup(r.content, 'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24c87d61-67f3-4465-9179-058185363b64",
   "metadata": {},
   "outputs": [],
   "source": [
    "href = soup.find(id='transcripts_table').tbody.td.a.attrs['href']\n",
    "transcript = re.sub(r'.*(ENST0\\d+)', r'\\1', href)\n",
    "print(href)\n",
    "print(transcript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9fef2af9-93d4-4b84-b41f-0e2afecb8522",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_new_transcript(req):\n",
    "    soup = BeautifulSoup(r.content, 'html.parser')\n",
    "    if soup.find(id='transcripts_table'):\n",
    "        href = soup.find(id='transcripts_table').tbody.td.a.attrs['href']\n",
    "        transcript = re.sub(r'.*(ENST0\\d+)', r'\\1', href)\n",
    "        print('   Current canonical transcript is', transcript)\n",
    "    else:\n",
    "        transcript = ''\n",
    "        print('   No current transcript found!')\n",
    "\n",
    "    return transcript\n",
    "\n",
    "def check_files_and_update_df(transcript):\n",
    "    bed_file_list = list(bed.glob(transcript + '*.bed'))\n",
    "    bed_files = len(bed_file_list)\n",
    "    fa_file_list = list(fa.glob(transcript + '*.fasta'))\n",
    "    fa_files = len(fa_file_list)\n",
    "\n",
    "    if bed_files == 1 and fa_files == 1:\n",
    "        print('   FA and BED files found. Updating dataframe with current information')\n",
    "        # update dataframe\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'EnsemblTranscriptID' ] = transcript\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed_files' ] = bed_files\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa_files' ] = fa_files\n",
    "    else:\n",
    "        print('   FA and BED file count invalid. File lists', bed_file_list, fa_file_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fe35c17e-5712-446e-97f7-1ccde66b5c5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing ENST00000263094\n",
      "   Current canonical transcript is ENST00000263094\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650372\n",
      "   Current canonical transcript is ENST00000650372\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647814\n",
      "   Current canonical transcript is ENST00000647814\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645237\n",
      "   Current canonical transcript is ENST00000645237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000405322\n",
      "   Current canonical transcript is ENST00000405322\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000238651\n",
      "   Current canonical transcript is ENST00000238651\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000646664\n",
      "   Current canonical transcript is ENST00000646664\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000366779\n",
      "   Current canonical transcript is ENST00000366779\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000649286\n",
      "   Current canonical transcript is ENST00000649286\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000644144\n",
      "   Current canonical transcript is ENST00000644144\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000642816\n",
      "   Current canonical transcript is ENST00000642816\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647789\n",
      "   Current canonical transcript is ENST00000647789\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000263440\n",
      "   Current canonical transcript is ENST00000263440\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647874\n",
      "   Current canonical transcript is ENST00000647874\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000374869\n",
      "   Current canonical transcript is ENST00000374869\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000535359\n",
      "   Current canonical transcript is ENST00000535359\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000643116\n",
      "   Current canonical transcript is ENST00000643116\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000542754\n",
      "   Current canonical transcript is ENST00000542754\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000649063\n",
      "   Current canonical transcript is ENST00000649063\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000606794\n",
      "   Current canonical transcript is ENST00000606794\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000564831\n",
      "   Current canonical transcript is ENST00000564831\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000397278\n",
      "   Current canonical transcript is ENST00000397278\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000272102\n",
      "   Current canonical transcript is ENST00000272102\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645722\n",
      "   Current canonical transcript is ENST00000645722\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000504924\n",
      "   Current canonical transcript is ENST00000504924\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000256496\n",
      "   Current canonical transcript is ENST00000256496\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000268314\n",
      "   Current canonical transcript is ENST00000268314\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000644486\n",
      "   Current canonical transcript is ENST00000644486\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000257749\n",
      "   Current canonical transcript is ENST00000257749\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000648566\n",
      "   Current canonical transcript is ENST00000648566\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000554140\n",
      "   Current canonical transcript is ENST00000554140\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000502613\n",
      "   Current canonical transcript is ENST00000502613\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000651561\n",
      "   Current canonical transcript is ENST00000651561\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000643237\n",
      "   Current canonical transcript is ENST00000643237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000421806\n",
      "   Current canonical transcript is ENST00000421806\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000534355\n",
      "   Current canonical transcript is ENST00000534355\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000397536\n",
      "   Current canonical transcript is ENST00000397536\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647956\n",
      "   Current canonical transcript is ENST00000647956\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000399562\n",
      "   Current canonical transcript is ENST00000399562\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371237\n",
      "   Current canonical transcript is ENST00000371237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371634\n",
      "   Current canonical transcript is ENST00000371634\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000382387\n",
      "   Current canonical transcript is ENST00000648587\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000649794\n",
      "   Current canonical transcript is ENST00000649794\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000394030\n",
      "   Current canonical transcript is ENST00000394030\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647358\n",
      "   Current canonical transcript is ENST00000647358\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000624406\n",
      "   Transcript is deprecated, resolved gene is ENSG00000274276\n",
      "   Current canonical transcript is ENST00000398168\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000442201\n",
      "   Transcript is deprecated, resolved gene is ENSG00000145075\n",
      "   Current canonical transcript is ENST00000471307\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000694881\n",
      "   Current canonical transcript is ENST00000694881\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000274026\n",
      "   Current canonical transcript is ENST00000274026\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000302014\n",
      "   Current canonical transcript is ENST00000302014\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000414716\n",
      "   Current canonical transcript is ENST00000414716\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000357172\n",
      "   Current canonical transcript is ENST00000357172\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000402394\n",
      "   Current canonical transcript is ENST00000402394\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000619537\n",
      "   Current canonical transcript is ENST00000619537\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000620528\n",
      "   Current canonical transcript is ENST00000620528\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000347519\n",
      "   Current canonical transcript is ENST00000347519\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000682950\n",
      "   Current canonical transcript is ENST00000682950\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000327325\n",
      "   Current canonical transcript is ENST00000327325\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000297268\n",
      "   Current canonical transcript is ENST00000297268\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000334504\n",
      "   Current canonical transcript is ENST00000334504\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000373157\n",
      "   Current canonical transcript is ENST00000373157\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647941\n",
      "   Current canonical transcript is ENST00000647941\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647684\n",
      "   Current canonical transcript is ENST00000647684\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000403662\n",
      "   Current canonical transcript is ENST00000403662\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000651323\n",
      "   Current canonical transcript is ENST00000651323\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650070\n",
      "   Current canonical transcript is ENST00000650070\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000488034\n",
      "   Current canonical transcript is ENST00000488034\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650725\n",
      "   Current canonical transcript is ENST00000650725\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000308775\n",
      "   Current canonical transcript is ENST00000308775\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000649689\n",
      "   Current canonical transcript is ENST00000649689\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000378278\n",
      "   Current canonical transcript is ENST00000378278\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000284031\n",
      "   Current canonical transcript is ENST00000284031\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000330503\n",
      "   Current canonical transcript is ENST00000330503\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000652689\n",
      "   Current canonical transcript is ENST00000652689\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645220\n",
      "   Current canonical transcript is ENST00000645220\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000649637\n",
      "   Current canonical transcript is ENST00000649637\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389054\n",
      "   Current canonical transcript is ENST00000389054\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389840\n",
      "   Current canonical transcript is ENST00000389840\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000648817\n",
      "   Current canonical transcript is ENST00000648817\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000520908\n",
      "   Current canonical transcript is ENST00000520908\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647188\n",
      "   Current canonical transcript is ENST00000647188\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000250454\n",
      "   Current canonical transcript is ENST00000250454\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000449606\n",
      "   Current canonical transcript is ENST00000449606\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000646607\n",
      "   Current canonical transcript is ENST00000646607\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000217182\n",
      "   Current canonical transcript is ENST00000217182\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000646441\n",
      "   Current canonical transcript is ENST00000646441\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000649764\n",
      "   Current canonical transcript is ENST00000649764\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000374450\n",
      "   Current canonical transcript is ENST00000374450\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000308167\n",
      "   Current canonical transcript is ENST00000308167\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000402918\n",
      "   Current canonical transcript is ENST00000402918\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647584\n",
      "   Current canonical transcript is ENST00000647584\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000647893\n",
      "   Current canonical transcript is ENST00000647893\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371506\n",
      "   Current canonical transcript is ENST00000371506\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389561\n",
      "   Current canonical transcript is ENST00000389561\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000261486\n",
      "   Current canonical transcript is ENST00000261486\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000260585\n",
      "   Current canonical transcript is ENST00000260585\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000288221\n",
      "   Current canonical transcript is ENST00000288221\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645284\n",
      "   Current canonical transcript is ENST00000645284\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000644823\n",
      "   Current canonical transcript is ENST00000644823\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000695508\n",
      "   Current canonical transcript is ENST00000695508\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000280987\n",
      "   Current canonical transcript is ENST00000280987\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000683666\n",
      "   Current canonical transcript is ENST00000683666\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368693\n",
      "   Current canonical transcript is ENST00000368693\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000297532\n",
      "   Current canonical transcript is ENST00000297532\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622495\n",
      "   Current canonical transcript is ENST00000622495\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000369384\n",
      "   Current canonical transcript is ENST00000453173\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000650385\n",
      "   Current canonical transcript is ENST00000650385\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000403106\n",
      "   Current canonical transcript is ENST00000403106\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000446176\n",
      "   Current canonical transcript is ENST00000446176\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645831\n",
      "   Current canonical transcript is ENST00000645831\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000648323\n",
      "   Current canonical transcript is ENST00000648323\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000380880\n",
      "   Current canonical transcript is ENST00000380880\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000645652\n",
      "   Current canonical transcript is ENST00000645652\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000512982\n",
      "   Current canonical transcript is ENST00000512982\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000393562\n",
      "   Current canonical transcript is ENST00000393562\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000311550\n",
      "   Current canonical transcript is ENST00000311550\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000648994\n",
      "   Current canonical transcript is ENST00000648994\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000491895\n",
      "   Current canonical transcript is ENST00000491895\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650454\n",
      "   Current canonical transcript is ENST00000650454\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000644452\n",
      "   Current canonical transcript is ENST00000644452\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000399440\n",
      "   Current canonical transcript is ENST00000399440\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000329125\n",
      "   Current canonical transcript is ENST00000329125\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000674370\n",
      "   Current canonical transcript is ENST00000674370\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368232\n",
      "   Current canonical transcript is ENST00000497287\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000419783\n",
      "   Current canonical transcript is ENST00000419783\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389614\n",
      "   Current canonical transcript is ENST00000389614\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000388825\n",
      "   Current canonical transcript is ENST00000388825\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000646743\n",
      "   Current canonical transcript is ENST00000646743\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000651619\n",
      "   Current canonical transcript is ENST00000651619\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650601\n",
      "   Current canonical transcript is ENST00000650601\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000337138\n",
      "   Current canonical transcript is ENST00000337138\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000361490\n",
      "   Current canonical transcript is ENST00000361490\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000510025\n",
      "   Current canonical transcript is ENST00000510025\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000372430\n",
      "   Current canonical transcript is ENST00000372430\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000649979\n",
      "   Current canonical transcript is ENST00000649979\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000650285\n",
      "   Current canonical transcript is ENST00000650285\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000695950\n",
      "   Current canonical transcript is ENST00000695950\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000447382\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[24], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprocessing\u001b[39m\u001b[38;5;124m'\u001b[39m, tid)\n\u001b[1;32m      3\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhttps://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m tid\n\u001b[0;32m----> 4\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      5\u001b[0m soup \u001b[38;5;241m=\u001b[39m BeautifulSoup(r\u001b[38;5;241m.\u001b[39mcontent, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mhtml.parser\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m re\u001b[38;5;241m.\u001b[39msearch(\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mThis transcript is not in the current gene set\u001b[39m\u001b[38;5;124m'\u001b[39m, soup\u001b[38;5;241m.\u001b[39mget_text()):\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/requests/api.py:73\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget\u001b[39m(url, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m     63\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     64\u001b[0m \n\u001b[1;32m     65\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m---> 73\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/requests/adapters.py:486\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    483\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    485\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 486\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    487\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    488\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    489\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    490\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    491\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    492\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    493\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    494\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    495\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    496\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    497\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    498\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    500\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    501\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/urllib3/connectionpool.py:791\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    788\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    790\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 791\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    803\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    806\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    807\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/urllib3/connectionpool.py:537\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 537\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    538\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    539\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m/scr/denobula/hawk/src/PTR-augur/.venv/lib64/python3.11/site-packages/urllib3/connection.py:461\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    458\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    460\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 461\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    463\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    464\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/usr/lib64/python3.11/http/client.py:1378\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1376\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1377\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1378\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1379\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[1;32m   1380\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib64/python3.11/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib64/python3.11/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfp\u001b[38;5;241m.\u001b[39mreadline(_MAXLINE \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1\u001b[39m), \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib64/python3.11/socket.py:706\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    704\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    705\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 706\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    707\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    708\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib64/python3.11/ssl.py:1311\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1307\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   1308\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1309\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[1;32m   1310\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[0;32m-> 1311\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1312\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1313\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[0;32m/usr/lib64/python3.11/ssl.py:1167\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[0;34m(self, len, buffer)\u001b[0m\n\u001b[1;32m   1165\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1166\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 1167\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1168\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1169\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for tid in df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ]: #.head(2):\n",
    "    print('processing', tid)\n",
    "    url = 'https://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=' + tid\n",
    "    r = requests.get(url)\n",
    "    soup = BeautifulSoup(r.content, 'html.parser')\n",
    "    if re.search(r'This transcript is not in the current gene set', soup.get_text()):\n",
    "        href = soup.td.next_sibling.a.attrs['href']\n",
    "        gene = re.sub(r'.*(ENSG0\\d+)', r'\\1', href)\n",
    "        print('   Transcript is deprecated, resolved gene is', gene)\n",
    "\n",
    "        url = 'https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=' + gene\n",
    "        r = requests.get(url)\n",
    "        transcript = find_new_transcript(r)\n",
    "        check_files_and_update_df(transcript)\n",
    "    elif re.search(r'Show transcript table', soup.get_text()):\n",
    "        transcript = find_new_transcript(r)\n",
    "        check_files_and_update_df(transcript)\n",
    "    else:\n",
    "        print('   Some other error occured for this transcript')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "834fbc22-ba69-4063-a639-cd7023e68465",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f910f9c-58c4-49a3-9466-bc734035ee89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write current pre processed table to file\n",
    "datafile = '../data/preproc.csv'\n",
    "df2.to_csv(datafile, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
