{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1beab998-5752-460c-88cc-811abeffc8e2",
   "metadata": {},
   "source": [
    "# Data Preprocessing\n",
    "\n",
    "## Overview\n",
    "\n",
    "The raw data from Eraslan et al. is a tab separated table containing gene names, related Ensembl IDs and measured or calculated values for mRNA abundance, protein abundance and protein-to-mRNA ratio.\n",
    "\n",
    "In the first Jupyter cells the data is roughly explored. After that the relevant values for the up coming analysis are extracted.\n",
    "\n",
    "Some problems arose:\n",
    "- Not all transcript IDs seem to be the current canonical form of transcript for a particular gene. In fact the given IDs point to transcripts that do not translate (in most cases). The gene needs to be identified in that case and the up to date transcript ID resolved.\n",
    "- BUT there are still some transcript IDs left that are annotate with \"nonsense mediated decay\". These will be thrown out as they do not successfully translate. \n",
    "- The Genecode data set contains duplicate files for some of the given transcript IDs. They can be easily filtered using a regex. The duplicates are the exact same files with different names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0fb525e9-e717-4129-99a3-89e5873b20e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# library dependencies\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e3b47d-618e-4b6c-9d5a-7ef18c8bbac4",
   "metadata": {},
   "source": [
    "## Reading the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8e5f0e17-2ae6-42ec-b039-eeab545d756e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw data file and path\n",
    "datafile = '../data/Eraslan-EV3.tsv'\n",
    "\n",
    "# sanity check if the file exists\n",
    "if not Path(datafile).is_file():\n",
    "    print('Data file not found!')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0edbe60a-daa9-419a-8635-184aeb0dda89",
   "metadata": {},
   "source": [
    "## Exploring the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1fc2f0e8-62f4-4b85-912e-50b04cc3d182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>EnsemblProteinID</th>\n",
       "      <th>Adrenal_mRNA</th>\n",
       "      <th>Appendices_mRNA</th>\n",
       "      <th>Brain_mRNA</th>\n",
       "      <th>Colon_mRNA</th>\n",
       "      <th>Duodenum_mRNA</th>\n",
       "      <th>Endometrium_mRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>ENSP00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.073</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.718</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.313</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>ENSP00000363105</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.971</td>\n",
       "      <td>2.324</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>5.147</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.202</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.8143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>ENSP00000323929</td>\n",
       "      <td>3.154</td>\n",
       "      <td>3.021</td>\n",
       "      <td>2.824</td>\n",
       "      <td>3.321</td>\n",
       "      <td>3.006</td>\n",
       "      <td>3.344</td>\n",
       "      <td>...</td>\n",
       "      <td>6.081</td>\n",
       "      <td>5.726</td>\n",
       "      <td>5.699</td>\n",
       "      <td>4.997</td>\n",
       "      <td>5.136</td>\n",
       "      <td>6.5349</td>\n",
       "      <td>5.820</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>ENSP00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.355</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274</td>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>ENSP00000384794</td>\n",
       "      <td>1.625</td>\n",
       "      <td>1.567</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>4.731</td>\n",
       "      <td>4.508</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0613</td>\n",
       "      <td>4.832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>ZXDB</td>\n",
       "      <td>ENSG00000198455</td>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>ENSP00000364023</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11571</th>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>ENSG00000162378</td>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>ENSP00000294353</td>\n",
       "      <td>1.930</td>\n",
       "      <td>1.589</td>\n",
       "      <td>1.995</td>\n",
       "      <td>1.627</td>\n",
       "      <td>1.531</td>\n",
       "      <td>2.082</td>\n",
       "      <td>...</td>\n",
       "      <td>4.962</td>\n",
       "      <td>4.987</td>\n",
       "      <td>5.076</td>\n",
       "      <td>4.827</td>\n",
       "      <td>4.255</td>\n",
       "      <td>4.0412</td>\n",
       "      <td>5.389</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11572</th>\n",
       "      <td>ZYX</td>\n",
       "      <td>ENSG00000159840</td>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>ENSP00000324422</td>\n",
       "      <td>2.414</td>\n",
       "      <td>2.978</td>\n",
       "      <td>2.349</td>\n",
       "      <td>2.257</td>\n",
       "      <td>2.572</td>\n",
       "      <td>3.175</td>\n",
       "      <td>...</td>\n",
       "      <td>6.268</td>\n",
       "      <td>5.564</td>\n",
       "      <td>5.708</td>\n",
       "      <td>6.284</td>\n",
       "      <td>6.159</td>\n",
       "      <td>5.8846</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>ZZEF1</td>\n",
       "      <td>ENSG00000074755</td>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>ENSP00000371051</td>\n",
       "      <td>1.851</td>\n",
       "      <td>1.904</td>\n",
       "      <td>1.866</td>\n",
       "      <td>2.140</td>\n",
       "      <td>2.175</td>\n",
       "      <td>1.689</td>\n",
       "      <td>...</td>\n",
       "      <td>5.540</td>\n",
       "      <td>5.181</td>\n",
       "      <td>5.303</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.110</td>\n",
       "      <td>5.0834</td>\n",
       "      <td>5.047</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>ENSG00000036549</td>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>ENSP00000359837</td>\n",
       "      <td>2.065</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1.940</td>\n",
       "      <td>1.908</td>\n",
       "      <td>2.079</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.458</td>\n",
       "      <td>3.380</td>\n",
       "      <td>3.010</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.4766</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11575 rows × 91 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneName    EnsemblGeneID EnsemblTranscriptID EnsemblProteinID  \\\n",
       "0         A1BG  ENSG00000121410     ENST00000263100  ENSP00000263100   \n",
       "1         A1CF  ENSG00000148584     ENST00000373993  ENSP00000363105   \n",
       "2          A2M  ENSG00000175899     ENST00000318602  ENSP00000323929   \n",
       "3        A2ML1  ENSG00000166535     ENST00000299698  ENSP00000299698   \n",
       "4       A4GALT  ENSG00000128274     ENST00000401850  ENSP00000384794   \n",
       "...        ...              ...                 ...              ...   \n",
       "11570     ZXDB  ENSG00000198455     ENST00000374888  ENSP00000364023   \n",
       "11571   ZYG11B  ENSG00000162378     ENST00000294353  ENSP00000294353   \n",
       "11572      ZYX  ENSG00000159840     ENST00000322764  ENSP00000324422   \n",
       "11573    ZZEF1  ENSG00000074755     ENST00000381638  ENSP00000371051   \n",
       "11574     ZZZ3  ENSG00000036549     ENST00000370801  ENSP00000359837   \n",
       "\n",
       "      Adrenal_mRNA Appendices_mRNA Brain_mRNA Colon_mRNA Duodenum_mRNA  \\\n",
       "0               NA           1.073         NA         NA            NA   \n",
       "1               NA              NA         NA      1.971         2.324   \n",
       "2            3.154           3.021      2.824      3.321         3.006   \n",
       "3               NA              NA      1.355         NA            NA   \n",
       "4            1.625           1.567         NA         NA            NA   \n",
       "...            ...             ...        ...        ...           ...   \n",
       "11570           NA              NA         NA         NA            NA   \n",
       "11571        1.930           1.589      1.995      1.627         1.531   \n",
       "11572        2.414           2.978      2.349      2.257         2.572   \n",
       "11573        1.851           1.904      1.866      2.140         2.175   \n",
       "11574        2.065              NA         NA      1.940         1.908   \n",
       "\n",
       "      Endometrium_mRNA  ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR  \\\n",
       "0                   NA  ...         NA             7.718                 NA   \n",
       "1                   NA  ...      5.147                NA              5.202   \n",
       "2                3.344  ...      6.081             5.726              5.699   \n",
       "3                   NA  ...         NA                NA                 NA   \n",
       "4                   NA  ...      4.731             4.508                 NA   \n",
       "...                ...  ...        ...               ...                ...   \n",
       "11570               NA  ...         NA                NA                 NA   \n",
       "11571            2.082  ...      4.962             4.987              5.076   \n",
       "11572            3.175  ...      6.268             5.564              5.708   \n",
       "11573            1.689  ...      5.540             5.181              5.303   \n",
       "11574            2.079  ...         NA             3.458              3.380   \n",
       "\n",
       "      Smoothmuscle_PTR Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR  \\\n",
       "0                   NA      7.313          NA         NA          NA   \n",
       "1                   NA         NA      5.8143         NA          NA   \n",
       "2                4.997      5.136      6.5349      5.820       6.060   \n",
       "3                   NA         NA          NA      2.350          NA   \n",
       "4                   NA         NA      4.0613      4.832          NA   \n",
       "...                ...        ...         ...        ...         ...   \n",
       "11570               NA         NA          NA         NA          NA   \n",
       "11571            4.827      4.255      4.0412      5.389       4.250   \n",
       "11572            6.284      6.159      5.8846      5.582       5.598   \n",
       "11573            5.038      5.110      5.0834      5.047       5.038   \n",
       "11574            3.010         NA      3.4766         NA          NA   \n",
       "\n",
       "      Tonsil_PTR Urinarybladder_PTR  \n",
       "0             NA                 NA  \n",
       "1             NA                 NA  \n",
       "2          5.675             5.8286  \n",
       "3          5.249                 NA  \n",
       "4             NA             4.2430  \n",
       "...          ...                ...  \n",
       "11570      4.681                 NA  \n",
       "11571      4.439             4.1460  \n",
       "11572      5.968             5.3358  \n",
       "11573      5.130             5.0619  \n",
       "11574      3.008             3.6041  \n",
       "\n",
       "[11575 rows x 91 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading the data into a dataframe and looking at the first entries\n",
    "df = pd.read_csv(datafile, sep='\\t')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10870254-b393-42b1-8a8d-7f38592b986b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 11575 entries, 0 to 11574\n",
      "Data columns (total 91 columns):\n",
      " #   Column                  Non-Null Count  Dtype \n",
      "---  ------                  --------------  ----- \n",
      " 0   GeneName                11575 non-null  object\n",
      " 1   EnsemblGeneID           11575 non-null  object\n",
      " 2   EnsemblTranscriptID     11575 non-null  object\n",
      " 3   EnsemblProteinID        11575 non-null  object\n",
      " 4   Adrenal_mRNA            11575 non-null  object\n",
      " 5   Appendices_mRNA         11575 non-null  object\n",
      " 6   Brain_mRNA              11575 non-null  object\n",
      " 7   Colon_mRNA              11575 non-null  object\n",
      " 8   Duodenum_mRNA           11575 non-null  object\n",
      " 9   Endometrium_mRNA        11575 non-null  object\n",
      " 10  Esophagus_mRNA          11575 non-null  object\n",
      " 11  Fallopiantube_mRNA      11575 non-null  object\n",
      " 12  Fat_mRNA                11575 non-null  object\n",
      " 13  Gallbladder_mRNA        11575 non-null  object\n",
      " 14  Heart_mRNA              11575 non-null  object\n",
      " 15  Kidney_mRNA             11575 non-null  object\n",
      " 16  Liver_mRNA              11575 non-null  object\n",
      " 17  Lung_mRNA               11575 non-null  object\n",
      " 18  Lymphnode_mRNA          11575 non-null  object\n",
      " 19  Ovary_mRNA              11575 non-null  object\n",
      " 20  Pancreas_mRNA           11575 non-null  object\n",
      " 21  Placenta_mRNA           11575 non-null  object\n",
      " 22  Prostate_mRNA           11575 non-null  object\n",
      " 23  Rectum_mRNA             11575 non-null  object\n",
      " 24  Salivarygland_mRNA      11575 non-null  object\n",
      " 25  Smallintestine_mRNA     11575 non-null  object\n",
      " 26  Smoothmuscle_mRNA       11575 non-null  object\n",
      " 27  Spleen_mRNA             11575 non-null  object\n",
      " 28  Stomach_mRNA            11575 non-null  object\n",
      " 29  Testis_mRNA             11575 non-null  object\n",
      " 30  Thyroid_mRNA            11575 non-null  object\n",
      " 31  Tonsil_mRNA             11575 non-null  object\n",
      " 32  Urinarybladder_mRNA     11575 non-null  object\n",
      " 33  Adrenal_protein         11575 non-null  object\n",
      " 34  Appendices_protein      11575 non-null  object\n",
      " 35  Brain_protein           11575 non-null  object\n",
      " 36  Colon_protein           11575 non-null  object\n",
      " 37  Duodenum_protein        11575 non-null  object\n",
      " 38  Endometrium_protein     11575 non-null  object\n",
      " 39  Esophagus_protein       11575 non-null  object\n",
      " 40  Fallopiantube_protein   11575 non-null  object\n",
      " 41  Fat_protein             11575 non-null  object\n",
      " 42  Gallbladder_protein     11575 non-null  object\n",
      " 43  Heart_protein           11575 non-null  object\n",
      " 44  Kidney_protein          11575 non-null  object\n",
      " 45  Liver_protein           11575 non-null  object\n",
      " 46  Lung_protein            11575 non-null  object\n",
      " 47  Lymphnode_protein       11575 non-null  object\n",
      " 48  Ovary_protein           11575 non-null  object\n",
      " 49  Pancreas_protein        11575 non-null  object\n",
      " 50  Placenta_protein        11575 non-null  object\n",
      " 51  Prostate_protein        11575 non-null  object\n",
      " 52  Rectum_protein          11575 non-null  object\n",
      " 53  Salivarygland_protein   11575 non-null  object\n",
      " 54  Smallintestine_protein  11575 non-null  object\n",
      " 55  Smoothmuscle_protein    11575 non-null  object\n",
      " 56  Spleen_protein          11575 non-null  object\n",
      " 57  Stomach_protein         11575 non-null  object\n",
      " 58  Testis_protein          11575 non-null  object\n",
      " 59  Thyroid_protein         11575 non-null  object\n",
      " 60  Tonsil_protein          11575 non-null  object\n",
      " 61  Urinarybladder_protein  11575 non-null  object\n",
      " 62  Adrenal_PTR             11575 non-null  object\n",
      " 63  Appendices_PTR          11575 non-null  object\n",
      " 64  Brain_PTR               11575 non-null  object\n",
      " 65  Colon_PTR               11575 non-null  object\n",
      " 66  Duodenum_PTR            11575 non-null  object\n",
      " 67  Endometrium_PTR         11575 non-null  object\n",
      " 68  Esophagus_PTR           11575 non-null  object\n",
      " 69  Fallopiantube_PTR       11575 non-null  object\n",
      " 70  Fat_PTR                 11575 non-null  object\n",
      " 71  Gallbladder_PTR         11575 non-null  object\n",
      " 72  Heart_PTR               11575 non-null  object\n",
      " 73  Kidney_PTR              11575 non-null  object\n",
      " 74  Liver_PTR               11575 non-null  object\n",
      " 75  Lung_PTR                11575 non-null  object\n",
      " 76  Lymphnode_PTR           11575 non-null  object\n",
      " 77  Ovary_PTR               11575 non-null  object\n",
      " 78  Pancreas_PTR            11575 non-null  object\n",
      " 79  Placenta_PTR            11575 non-null  object\n",
      " 80  Prostate_PTR            11575 non-null  object\n",
      " 81  Rectum_PTR              11575 non-null  object\n",
      " 82  Salivarygland_PTR       11575 non-null  object\n",
      " 83  Smallintestine_PTR      11575 non-null  object\n",
      " 84  Smoothmuscle_PTR        11575 non-null  object\n",
      " 85  Spleen_PTR              11575 non-null  object\n",
      " 86  Stomach_PTR             11575 non-null  object\n",
      " 87  Testis_PTR              11575 non-null  object\n",
      " 88  Thyroid_PTR             11575 non-null  object\n",
      " 89  Tonsil_PTR              11575 non-null  object\n",
      " 90  Urinarybladder_PTR      11575 non-null  object\n",
      "dtypes: object(91)\n",
      "memory usage: 8.0+ MB\n"
     ]
    }
   ],
   "source": [
    "# looking at the columns\n",
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0ee1e23d-728a-4243-8050-f6b999fa2cbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneName</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>EnsemblProteinID</th>\n",
       "      <th>Adrenal_mRNA</th>\n",
       "      <th>Appendices_mRNA</th>\n",
       "      <th>Brain_mRNA</th>\n",
       "      <th>Colon_mRNA</th>\n",
       "      <th>Duodenum_mRNA</th>\n",
       "      <th>Endometrium_mRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>...</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>11575</td>\n",
       "      <td>2132</td>\n",
       "      <td>2049</td>\n",
       "      <td>2020</td>\n",
       "      <td>2143</td>\n",
       "      <td>2183</td>\n",
       "      <td>2043</td>\n",
       "      <td>...</td>\n",
       "      <td>3175</td>\n",
       "      <td>3297</td>\n",
       "      <td>3117</td>\n",
       "      <td>3248</td>\n",
       "      <td>3109</td>\n",
       "      <td>7372</td>\n",
       "      <td>3299</td>\n",
       "      <td>3238</td>\n",
       "      <td>3102</td>\n",
       "      <td>6971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>ENSP00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3162</td>\n",
       "      <td>4011</td>\n",
       "      <td>3079</td>\n",
       "      <td>3396</td>\n",
       "      <td>2971</td>\n",
       "      <td>3395</td>\n",
       "      <td>...</td>\n",
       "      <td>3471</td>\n",
       "      <td>3643</td>\n",
       "      <td>2981</td>\n",
       "      <td>3632</td>\n",
       "      <td>4022</td>\n",
       "      <td>3124</td>\n",
       "      <td>2938</td>\n",
       "      <td>3962</td>\n",
       "      <td>3603</td>\n",
       "      <td>3697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 91 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       GeneName    EnsemblGeneID EnsemblTranscriptID EnsemblProteinID  \\\n",
       "count     11575            11575               11575            11575   \n",
       "unique    11575            11575               11575            11575   \n",
       "top        A1BG  ENSG00000121410     ENST00000263100  ENSP00000263100   \n",
       "freq          1                1                   1                1   \n",
       "\n",
       "       Adrenal_mRNA Appendices_mRNA Brain_mRNA Colon_mRNA Duodenum_mRNA  \\\n",
       "count         11575           11575      11575      11575         11575   \n",
       "unique         2132            2049       2020       2143          2183   \n",
       "top              NA              NA         NA         NA            NA   \n",
       "freq           3162            4011       3079       3396          2971   \n",
       "\n",
       "       Endometrium_mRNA  ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR  \\\n",
       "count             11575  ...      11575             11575              11575   \n",
       "unique             2043  ...       3175              3297               3117   \n",
       "top                  NA  ...         NA                NA                 NA   \n",
       "freq               3395  ...       3471              3643               2981   \n",
       "\n",
       "       Smoothmuscle_PTR Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR  \\\n",
       "count             11575      11575       11575      11575       11575   \n",
       "unique             3248       3109        7372       3299        3238   \n",
       "top                  NA         NA          NA         NA          NA   \n",
       "freq               3632       4022        3124       2938        3962   \n",
       "\n",
       "       Tonsil_PTR Urinarybladder_PTR  \n",
       "count       11575              11575  \n",
       "unique       3102               6971  \n",
       "top            NA                 NA  \n",
       "freq         3603               3697  \n",
       "\n",
       "[4 rows x 91 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de10a03-6d3e-4348-a970-1394d58ef4c6",
   "metadata": {},
   "source": [
    "Between ~3000 and ~4000 values in each of the 11575 rows are NA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "049d92a4-c181-43b9-81db-e2fbd4a7a9cd",
   "metadata": {},
   "source": [
    "## Extracting the relevant columns\n",
    "\n",
    "Only the _EnsemblTranscriptID_ and _PTR_ values per tissue are necessary for training the network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5a7b83b9-4921-422f-b99e-622e830e56d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Rectum_PTR</th>\n",
       "      <th>Salivarygland_PTR</th>\n",
       "      <th>Smallintestine_PTR</th>\n",
       "      <th>Smoothmuscle_PTR</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.277</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.841</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.718</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.313</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.135</td>\n",
       "      <td>5.371</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>5.147</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.202</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.8143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>6.290</td>\n",
       "      <td>6.328</td>\n",
       "      <td>5.948</td>\n",
       "      <td>5.811</td>\n",
       "      <td>6.068</td>\n",
       "      <td>5.383</td>\n",
       "      <td>5.881</td>\n",
       "      <td>6.119</td>\n",
       "      <td>6.410</td>\n",
       "      <td>...</td>\n",
       "      <td>6.081</td>\n",
       "      <td>5.726</td>\n",
       "      <td>5.699</td>\n",
       "      <td>4.997</td>\n",
       "      <td>5.136</td>\n",
       "      <td>6.5349</td>\n",
       "      <td>5.820</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.995</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.129</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>3.843</td>\n",
       "      <td>4.601</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.013</td>\n",
       "      <td>3.683</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>4.731</td>\n",
       "      <td>4.508</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0613</td>\n",
       "      <td>4.832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11571</th>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>4.461</td>\n",
       "      <td>5.013</td>\n",
       "      <td>5.047</td>\n",
       "      <td>4.566</td>\n",
       "      <td>5.184</td>\n",
       "      <td>4.826</td>\n",
       "      <td>5.102</td>\n",
       "      <td>4.670</td>\n",
       "      <td>5.756</td>\n",
       "      <td>...</td>\n",
       "      <td>4.962</td>\n",
       "      <td>4.987</td>\n",
       "      <td>5.076</td>\n",
       "      <td>4.827</td>\n",
       "      <td>4.255</td>\n",
       "      <td>4.0412</td>\n",
       "      <td>5.389</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11572</th>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.524</td>\n",
       "      <td>5.478</td>\n",
       "      <td>5.915</td>\n",
       "      <td>5.811</td>\n",
       "      <td>5.817</td>\n",
       "      <td>5.943</td>\n",
       "      <td>5.509</td>\n",
       "      <td>4.931</td>\n",
       "      <td>...</td>\n",
       "      <td>6.268</td>\n",
       "      <td>5.564</td>\n",
       "      <td>5.708</td>\n",
       "      <td>6.284</td>\n",
       "      <td>6.159</td>\n",
       "      <td>5.8846</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>5.112</td>\n",
       "      <td>4.918</td>\n",
       "      <td>5.139</td>\n",
       "      <td>5.190</td>\n",
       "      <td>5.442</td>\n",
       "      <td>5.602</td>\n",
       "      <td>4.715</td>\n",
       "      <td>4.956</td>\n",
       "      <td>5.033</td>\n",
       "      <td>...</td>\n",
       "      <td>5.540</td>\n",
       "      <td>5.181</td>\n",
       "      <td>5.303</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.110</td>\n",
       "      <td>5.0834</td>\n",
       "      <td>5.047</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>3.165</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.539</td>\n",
       "      <td>3.629</td>\n",
       "      <td>3.616</td>\n",
       "      <td>4.203</td>\n",
       "      <td>3.400</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.458</td>\n",
       "      <td>3.380</td>\n",
       "      <td>3.010</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.4766</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11575 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EnsemblTranscriptID Adrenal_PTR Appendices_PTR Brain_PTR Colon_PTR  \\\n",
       "0         ENST00000263100          NA          8.277        NA        NA   \n",
       "1         ENST00000373993          NA             NA        NA     5.135   \n",
       "2         ENST00000318602       6.290          6.328     5.948     5.811   \n",
       "3         ENST00000299698          NA             NA     3.995        NA   \n",
       "4         ENST00000401850       3.843          4.601        NA        NA   \n",
       "...                   ...         ...            ...       ...       ...   \n",
       "11570     ENST00000374888          NA             NA        NA        NA   \n",
       "11571     ENST00000294353       4.461          5.013     5.047     4.566   \n",
       "11572     ENST00000322764       5.664          5.524     5.478     5.915   \n",
       "11573     ENST00000381638       5.112          4.918     5.139     5.190   \n",
       "11574     ENST00000370801       3.165             NA        NA     3.539   \n",
       "\n",
       "      Duodenum_PTR Endometrium_PTR Esophagus_PTR Fallopiantube_PTR Fat_PTR  \\\n",
       "0               NA              NA            NA             7.841      NA   \n",
       "1            5.371              NA            NA                NA      NA   \n",
       "2            6.068           5.383         5.881             6.119   6.410   \n",
       "3               NA              NA         4.129                NA      NA   \n",
       "4               NA              NA         4.013             3.683      NA   \n",
       "...            ...             ...           ...               ...     ...   \n",
       "11570           NA              NA            NA                NA      NA   \n",
       "11571        5.184           4.826         5.102             4.670   5.756   \n",
       "11572        5.811           5.817         5.943             5.509   4.931   \n",
       "11573        5.442           5.602         4.715             4.956   5.033   \n",
       "11574        3.629           3.616         4.203             3.400      NA   \n",
       "\n",
       "       ... Rectum_PTR Salivarygland_PTR Smallintestine_PTR Smoothmuscle_PTR  \\\n",
       "0      ...         NA             7.718                 NA               NA   \n",
       "1      ...      5.147                NA              5.202               NA   \n",
       "2      ...      6.081             5.726              5.699            4.997   \n",
       "3      ...         NA                NA                 NA               NA   \n",
       "4      ...      4.731             4.508                 NA               NA   \n",
       "...    ...        ...               ...                ...              ...   \n",
       "11570  ...         NA                NA                 NA               NA   \n",
       "11571  ...      4.962             4.987              5.076            4.827   \n",
       "11572  ...      6.268             5.564              5.708            6.284   \n",
       "11573  ...      5.540             5.181              5.303            5.038   \n",
       "11574  ...         NA             3.458              3.380            3.010   \n",
       "\n",
       "      Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR Tonsil_PTR  \\\n",
       "0          7.313          NA         NA          NA         NA   \n",
       "1             NA      5.8143         NA          NA         NA   \n",
       "2          5.136      6.5349      5.820       6.060      5.675   \n",
       "3             NA          NA      2.350          NA      5.249   \n",
       "4             NA      4.0613      4.832          NA         NA   \n",
       "...          ...         ...        ...         ...        ...   \n",
       "11570         NA          NA         NA          NA      4.681   \n",
       "11571      4.255      4.0412      5.389       4.250      4.439   \n",
       "11572      6.159      5.8846      5.582       5.598      5.968   \n",
       "11573      5.110      5.0834      5.047       5.038      5.130   \n",
       "11574         NA      3.4766         NA          NA      3.008   \n",
       "\n",
       "      Urinarybladder_PTR  \n",
       "0                     NA  \n",
       "1                     NA  \n",
       "2                 5.8286  \n",
       "3                     NA  \n",
       "4                 4.2430  \n",
       "...                  ...  \n",
       "11570                 NA  \n",
       "11571             4.1460  \n",
       "11572             5.3358  \n",
       "11573             5.0619  \n",
       "11574             3.6041  \n",
       "\n",
       "[11575 rows x 30 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df[['EnsemblTranscriptID'] + [ col for col in df.columns if col.endswith('_PTR') ]].copy()\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76fe860-fd42-43ec-9818-0ffab352701e",
   "metadata": {},
   "source": [
    "Cross referencing the transcript IDs with BED and Fasta files from the gencode data set (43).\n",
    "\n",
    "The path set below expects the gencode repo to be relative to this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5acf60f0-d6e7-4955-b214-23454e27a919",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000435683 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263817 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370449 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000376887 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000260645 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622407 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000331789 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000366779 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355413 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000373176 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000313871 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000313871.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000313871.9:0-3199.fasta\n",
      "ENST00000564546 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000374855 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371108 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000319144 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330258 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620652 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355272 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622409 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000348624 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611099 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000431282 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000422706 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000540651 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378004 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620747 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611208 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000538189 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381317 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381317.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381317.9:0-2065.fasta\n",
      "ENST00000617719 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000537989 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370580 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000443071 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000335016 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371265 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000383778 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000286067 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622257 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000559313 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000542285 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000333059 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000535057 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000224181 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000382387 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309893 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000264122 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000270279 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000624406 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000442201 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392347 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618014 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000401743 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381192 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381192.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381192.10:0-1129.fasta\n",
      "ENST00000453495 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615636 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000260595 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619537 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620528 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000609024 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619275 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000248228 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620463 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545689 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265379 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000264193 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000215939 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000432318 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000432318.8.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000432318.8:0-1940.fasta\n",
      "ENST00000262825 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000315684 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372621 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611509 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409414 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545947 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000361951 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396817 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000539042 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000507955 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000397239 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409775 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000266991 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000334651 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000334651.11.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000334651.11:0-2579.fasta\n",
      "ENST00000398557 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000585328 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000342382 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000256935 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000339659 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618169 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000563459 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000334286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000298049 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000245323 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000269349 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621148 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615377 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613079 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000251101 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000360971 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000472938 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000389562 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621003 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613142 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612797 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000377828 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556177 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620555 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396472 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622754 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368695 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000540185 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000427718 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000369384 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613418 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000302053 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000443566 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000380874 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000422223 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000310160 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000505428 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621232 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000628124 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000568107 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622544 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000229416 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000322315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620807 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611961 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000409774 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368232 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620890 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612794 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622181 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000251808 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000216951 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000326153 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000326153.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000326153.10:0-1875.fasta\n",
      "ENST00000380299 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000557538 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613394 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000256216 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372425 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263642 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558762 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000389983 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000412504 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381469 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381469.7.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381469.7:0-1476.fasta\n",
      "ENST00000401393 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000361976 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396435 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000379389 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000287497 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622834 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378565 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000543780 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368089 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000262888 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000406396 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000355193 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265023 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614544 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612396 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000415258 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000592991 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541779 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000311322 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000263816 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000500728 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000418342 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000395528 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000215739 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000348428 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000226578 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000493901 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000523944 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000629399 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000599213 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000434178 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000537344 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000396617 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000591601 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000281317 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000233616 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000380172 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000216605 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000625889 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000359786 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265944 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541942 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368063 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000331837 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378062 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000411638 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000627621 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000371527 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622383 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000436936 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000344089 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000277541 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620134 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000405342 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000376214 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381297 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381297.10.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381297.10:0-4267.fasta\n",
      "ENST00000379806 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000542749 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000370184 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000372389 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614426 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000404875 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000303531 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615510 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000357304 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000632805 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000606149 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000620216 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000619580 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000306726 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000403432 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000527788 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000329286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610732 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000343115 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000625761 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611665 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617334 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617441 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556710 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000318842 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000424325 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000464167 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368567 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534062 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000590320 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617721 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611159 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000621965 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000377475 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000541726 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611680 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611222 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615547 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616787 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000612212 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367104 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616363 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000330479 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000265735 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000357436 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534261 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558405 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000298472 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000397260 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000273158 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381401 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381401.11.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381401.11:0-1451.fasta\n",
      "ENST00000427851 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392845 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000424970 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000539553 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000400286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614072 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000602728 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000556626 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000342756 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000586243 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000267540 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000378654 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000538427 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615887 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000432496 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616883 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000609199 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000611869 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000536810 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000535286 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000613071 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000615733 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000558083 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392129 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610315 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000617947 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000588847 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000392451 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000406386 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000290846 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000454289 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000622375 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000368563 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000429002 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000610664 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367924 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000367274 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000614096 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000616789 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000286448 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000286448.12.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000286448.12:0-2594.fasta\n",
      "ENST00000261517 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000369130 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000618527 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000434753 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000400602 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000381223 more than one BED/Fasta file present. selecting\n",
      "    ../../GENCODE43/protein_coding/BED6__protein_coding_strict/ENST00000381223.9.bed\n",
      "    ../../GENCODE43/protein_coding/FA_protein_coding_strict_mRNA/ENST00000381223.9:0-4510.fasta\n",
      "ENST00000373452 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000545588 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309776 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000534834 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000309495 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "ENST00000543588 bed count: 0 fa count: 0 bed files: [] fa files: []\n",
      "searched for 11575 and found 11281\n",
      "found multiple files for 11 transcripts\n",
      "missing or otherwise off: 294\n"
     ]
    }
   ],
   "source": [
    "# raw data paths\n",
    "gencode_path = '../../GENCODE43/protein_coding/'\n",
    "bed = Path(gencode_path) / 'BED6__protein_coding_strict/'\n",
    "fa = Path(gencode_path) / 'FA_protein_coding_strict_mRNA/'\n",
    "\n",
    "# file names look like this\n",
    "# for the BED file : ENST00000370801.8.bed\n",
    "# for the Fasta file : ENST00000370801.8:0-6412.fasta\n",
    "# .8 denotes the current Ensemble version\n",
    "# :0-6412 is the nucleotide length\n",
    "\n",
    "# count of processed transcript IDs\n",
    "count_all = 0\n",
    "# success count\n",
    "count_found = 0\n",
    "# multiple files found for transcript\n",
    "count_multi = 0\n",
    "\n",
    "# extend the dataframe\n",
    "# number of files found\n",
    "df2['bed_files'] = 0\n",
    "df2['fa_files'] = 0\n",
    "# file path and name\n",
    "df2['bed'] = ''\n",
    "df2['fa'] = ''\n",
    "\n",
    "# checking if all the transcript Fasta and BED files per transcript exist\n",
    "for tid in df2['EnsemblTranscriptID']:\n",
    "    # inclrease over all count\n",
    "    count_all += 1\n",
    "\n",
    "    # list and count files\n",
    "    bed_file_list = list(bed.glob(tid + '*.bed'))\n",
    "    bed_file_count = len(bed_file_list)\n",
    "    fa_file_list = list(fa.glob(tid + '*.fasta'))\n",
    "    fa_file_count = len(fa_file_list)\n",
    "\n",
    "    # update dataframe\n",
    "    df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed_files'] = bed_file_count\n",
    "    df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa_files'] = fa_file_count\n",
    "\n",
    "    # check BED and Fasta file count\n",
    "    if bed_file_count == 1 and fa_file_count == 1:\n",
    "        # exctly one BED and FA file\n",
    "        \n",
    "        # increase hit count\n",
    "        count_found += 1\n",
    "        \n",
    "        # update file name information\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed'] = str(bed_file_list[0])\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa'] = str(fa_file_list[0])\n",
    "    elif bed_file_count == 2 and fa_file_count == 2:\n",
    "        # special case where there are duplicate files\n",
    "        print(tid, 'more than one BED/Fasta file present. selecting')\n",
    "\n",
    "        # increase hit count\n",
    "        count_found += 1\n",
    "        count_multi += 1\n",
    "\n",
    "        # find correct BED file and update table\n",
    "        for f in bed_file_list:\n",
    "            temp_bed_file = str(f)\n",
    "            if re.search(r'.*ENST\\d+\\.\\d+.bed', temp_bed_file):\n",
    "                df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed'] = temp_bed_file\n",
    "                print('   ', temp_bed_file)\n",
    "\n",
    "        # find correct Fasta file and update table\n",
    "        for f in fa_file_list:\n",
    "            temp_fa_file = str(f)\n",
    "            if re.search(r'.*ENST\\d+\\.\\d+:\\d+-\\d+.fasta', temp_fa_file):\n",
    "                df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa'] = temp_fa_file\n",
    "                print('   ', temp_fa_file)\n",
    "\n",
    "        # update file count in table\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'bed_files'] = 1\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == tid, 'fa_files'] = 1\n",
    "    else:\n",
    "        # everything else ends up here\n",
    "        print(tid, 'bed count:', bed_file_count, 'fa count:', fa_file_count, 'bed files:', bed_file_list, 'fa files:', fa_file_list)\n",
    "\n",
    "print('searched for', count_all, 'and found', count_found)\n",
    "print('found multiple files for', count_multi, 'transcripts')\n",
    "print('missing or otherwise off:', count_all - count_found)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "33e20baf-8127-415b-a417-2f519fd2557d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# temporaray backup of df2\n",
    "# for restoration if the processing below fails (again)\n",
    "df2_backup = df2.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7a9ae935-7fe4-472c-b6be-916bb9d9209d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ONLY RUN THIS FOR DEBUGGING\n",
    "df2 = df2_backup.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ba103ec0-a297-43fe-a45d-ba6d1b8b493e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "      <th>bed_files</th>\n",
       "      <th>fa_files</th>\n",
       "      <th>bed</th>\n",
       "      <th>fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [EnsemblTranscriptID, Adrenal_PTR, Appendices_PTR, Brain_PTR, Colon_PTR, Duodenum_PTR, Endometrium_PTR, Esophagus_PTR, Fallopiantube_PTR, Fat_PTR, Gallbladder_PTR, Heart_PTR, Kidney_PTR, Liver_PTR, Lung_PTR, Lymphnode_PTR, Ovary_PTR, Pancreas_PTR, Placenta_PTR, Prostate_PTR, Rectum_PTR, Salivarygland_PTR, Smallintestine_PTR, Smoothmuscle_PTR, Spleen_PTR, Stomach_PTR, Testis_PTR, Thyroid_PTR, Tonsil_PTR, Urinarybladder_PTR, bed_files, fa_files, bed, fa]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 34 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# entries with two transcript files per entry\n",
    "# the gencod data set contains a couple of transcript files with multiple different names\n",
    "# file countent is exaclty the same\n",
    "# this has been corrected in the previous cell so this sanity check should reveal 0 rows\n",
    "df2.loc[ df2['bed_files'] == 2 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "53cd8bba-8966-4f95-98c4-b5942e0b07be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "294"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check again the number of missing files (with a count of 0 in the bed_file column)\n",
    "df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ea1de33f-13c7-4f12-b6fe-24e0fc3319b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "294"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[ df2['bed'] == '' ]['EnsemblTranscriptID' ].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7cda9f82-6566-4955-b960-c675b07cbf62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28       ENST00000435683\n",
       "33       ENST00000263817\n",
       "39       ENST00000370449\n",
       "41       ENST00000376887\n",
       "54       ENST00000260645\n",
       "              ...       \n",
       "11323    ENST00000545588\n",
       "11411    ENST00000309776\n",
       "11441    ENST00000534834\n",
       "11513    ENST00000309495\n",
       "11559    ENST00000543588\n",
       "Name: EnsemblTranscriptID, Length: 294, dtype: object"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show some of the IDs with missing files\n",
    "df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a232ea0-92dd-453c-9089-c1fda02ac35d",
   "metadata": {},
   "source": [
    "Sampling the IDs for which files are missing showed that the transcript is either deprecated or not the canonical form in the gencode data set (43). For some IDs the transcripts either do not translate or a different gene translates to the associated protein.\n",
    "\n",
    "This is a natural evolution since the Eraslan et al. research took place 2019 the underlying data in the Ensembl database got updated with current research results.\n",
    "\n",
    "The following cells query the Ensembl web site directly for the transcript IDs in question as it's the fastest way to resolve this issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9fef2af9-93d4-4b84-b41f-0e2afecb8522",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_new_transcript(content):\n",
    "    \"\"\"Method to extract a specific transcript ID from an HTML document.\n",
    "\n",
    "    Keyword Arguments:\n",
    "    req -- Python request object content\n",
    "\n",
    "    Returns:\n",
    "    A string either empty or containing the transcript ID.\n",
    "    \"\"\"\n",
    "    # parse the HTML document\n",
    "    soup = BeautifulSoup(content, 'html.parser')\n",
    "    # check if a specific table exists\n",
    "    if soup.find(id='transcripts_table'):\n",
    "        # if so extract the transcript ID\n",
    "        href = soup.find(id='transcripts_table').tbody.td.a.attrs['href']\n",
    "        transcript = re.sub(r'.*(ENST0\\d+)', r'\\1', href)\n",
    "        print('   Current canonical transcript is', transcript)\n",
    "    else:\n",
    "        # if not return an empty string\n",
    "        transcript = ''\n",
    "        print('   No current transcript found!')\n",
    "\n",
    "    return transcript\n",
    "\n",
    "def check_files_and_update_df(old_tid, new_tid):\n",
    "    \"\"\"Cross reference the transcript ID with files in the gencode data set\n",
    "    (bad hack as it uses variables globally defined at the beginning of this notebook!)\n",
    "\n",
    "    Keyword Arguments:\n",
    "    transcript -- the transcript ID string\n",
    "\n",
    "    Returns:\n",
    "    True if files were found for a transcript ID and False otherwise\n",
    "    \"\"\"\n",
    "    # search and count files with a given name\n",
    "    bed_file_list = list(bed.glob(new_tid + '*.bed'))\n",
    "    bed_files = len(bed_file_list)\n",
    "    fa_file_list = list(fa.glob(new_tid + '*.fasta'))\n",
    "    fa_files = len(fa_file_list)\n",
    "\n",
    "    # check how many files were found\n",
    "    if bed_files == 1 and fa_files == 1:\n",
    "        # if it's 1 everything is perfect\n",
    "        print('   FA and BED files found. Updating dataframe with current information')\n",
    "        # update dataframe\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'bed_files' ] = bed_files\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'fa_files' ] = fa_files\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'bed'] = str(bed_file_list[0])\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'fa'] = str(fa_file_list[0])\n",
    "        df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'EnsemblTranscriptID' ] = new_tid\n",
    "        \n",
    "        return True\n",
    "    else:\n",
    "        # if there are many manual processing is needed\n",
    "        print('   FA and BED file count invalid. File lists', bed_file_list, fa_file_list)\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "fe35c17e-5712-446e-97f7-1ccde66b5c5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing ENST00000263817\n",
      "   Transcript is deprecated, resolved gene is ENSG00000073734\n",
      "   Current canonical transcript is ENST00000650372\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000370449\n",
      "   Transcript is deprecated, resolved gene is ENSG00000023839\n",
      "   Current canonical transcript is ENST00000647814\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000376887\n",
      "   Transcript is deprecated, resolved gene is ENSG00000125257\n",
      "   Current canonical transcript is ENST00000645237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000260645\n",
      "   Transcript is deprecated, resolved gene is ENSG00000138075\n",
      "   Current canonical transcript is ENST00000405322\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622407\n",
      "   Transcript is deprecated, resolved gene is ENSG00000119673\n",
      "   Current canonical transcript is ENST00000238651\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000331789\n",
      "   Transcript is deprecated, resolved gene is ENSG00000075624\n",
      "   Current canonical transcript is ENST00000646664\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000366779\n",
      "   Current canonical transcript is ENST00000366779\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000355413\n",
      "   Transcript is deprecated, resolved gene is ENSG00000006530\n",
      "   Current canonical transcript is ENST00000649286\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000373176\n",
      "   Transcript is deprecated, resolved gene is ENSG00000106992\n",
      "   Current canonical transcript is ENST00000644144\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000564546\n",
      "   Transcript is deprecated, resolved gene is ENSG00000149925\n",
      "   Current canonical transcript is ENST00000642816\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000374855\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136872\n",
      "   Current canonical transcript is ENST00000647789\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371108\n",
      "   Transcript is deprecated, resolved gene is ENSG00000088035\n",
      "   Current canonical transcript is ENST00000263440\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000319144\n",
      "   Transcript is deprecated, resolved gene is ENSG00000179477\n",
      "   Current canonical transcript is ENST00000647874\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000330258\n",
      "   Transcript is deprecated, resolved gene is ENSG00000184675\n",
      "   Current canonical transcript is ENST00000374869\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620652\n",
      "   Transcript is deprecated, resolved gene is ENSG00000103723\n",
      "   Current canonical transcript is ENST00000535359\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000355272\n",
      "   Transcript is deprecated, resolved gene is ENSG00000065000\n",
      "   Current canonical transcript is ENST00000643116\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622409\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100478\n",
      "   Current canonical transcript is ENST00000542754\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000348624\n",
      "   Transcript is deprecated, resolved gene is ENSG00000242802\n",
      "   Current canonical transcript is ENST00000649063\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611099\n",
      "   Transcript is deprecated, resolved gene is ENSG00000134817\n",
      "   Current canonical transcript is ENST00000606794\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000431282\n",
      "   Transcript is deprecated, resolved gene is ENSG00000184730\n",
      "   Current canonical transcript is ENST00000564831\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000422706\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100342\n",
      "   Current canonical transcript is ENST00000397278\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000540651\n",
      "   Transcript is deprecated, resolved gene is ENSG00000143761\n",
      "   Current canonical transcript is ENST00000272102\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000378004\n",
      "   Transcript is deprecated, resolved gene is ENSG00000145819\n",
      "   Current canonical transcript is ENST00000645722\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620747\n",
      "   Transcript is deprecated, resolved gene is ENSG00000185305\n",
      "   Current canonical transcript is ENST00000504924\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611208\n",
      "   Transcript is deprecated, resolved gene is ENSG00000134108\n",
      "   Current canonical transcript is ENST00000256496\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000538189\n",
      "   Transcript is deprecated, resolved gene is ENSG00000140691\n",
      "   Current canonical transcript is ENST00000268314\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000617719\n",
      "   Transcript is deprecated, resolved gene is ENSG00000066427\n",
      "   Current canonical transcript is ENST00000644486\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000537989\n",
      "   Transcript is deprecated, resolved gene is ENSG00000112182\n",
      "   Current canonical transcript is ENST00000257749\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000370580\n",
      "   Transcript is deprecated, resolved gene is ENSG00000142867\n",
      "   Current canonical transcript is ENST00000648566\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000443071\n",
      "   Transcript is deprecated, resolved gene is ENSG00000183092\n",
      "   Current canonical transcript is ENST00000554140\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000335016\n",
      "   Transcript is deprecated, resolved gene is ENSG00000138756\n",
      "   Current canonical transcript is ENST00000502613\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371265\n",
      "   Transcript is deprecated, resolved gene is ENSG00000162399\n",
      "   Current canonical transcript is ENST00000651561\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000383778\n",
      "   Transcript is deprecated, resolved gene is ENSG00000169814\n",
      "   Current canonical transcript is ENST00000643237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000286067\n",
      "   Transcript is deprecated, resolved gene is ENSG00000196233\n",
      "   Current canonical transcript is ENST00000421806\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622257\n",
      "   Transcript is deprecated, resolved gene is ENSG00000211450\n",
      "   Current canonical transcript is ENST00000534355\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000559313\n",
      "   Current canonical transcript is ENST00000397536\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000542285\n",
      "   Transcript is deprecated, resolved gene is ENSG00000159403\n",
      "   Current canonical transcript is ENST00000647956\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000333059\n",
      "   Transcript is deprecated, resolved gene is ENSG00000242259\n",
      "   Current canonical transcript is ENST00000399562\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000535057\n",
      "   Transcript is deprecated, resolved gene is ENSG00000021852\n",
      "   Current canonical transcript is ENST00000371237\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000224181\n",
      "   Transcript is deprecated, resolved gene is ENSG00000176919\n",
      "   Current canonical transcript is ENST00000371634\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000382387\n",
      "   Current canonical transcript is ENST00000648587\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000309893\n",
      "   Transcript is deprecated, resolved gene is ENSG00000174990\n",
      "   Current canonical transcript is ENST00000649794\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000264122\n",
      "   Transcript is deprecated, resolved gene is ENSG00000114423\n",
      "   Current canonical transcript is ENST00000394030\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000270279\n",
      "   Transcript is deprecated, resolved gene is ENSG00000142273\n",
      "   Current canonical transcript is ENST00000647358\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000624406\n",
      "   Transcript is deprecated, resolved gene is ENSG00000274276\n",
      "   Current canonical transcript is ENST00000398168\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000442201\n",
      "   Transcript is deprecated, resolved gene is ENSG00000145075\n",
      "   Current canonical transcript is ENST00000471307\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "   Transcript is deprecated, resolved gene is ENSG00000284862\n",
      "   Current canonical transcript is ENST00000476379\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000392347\n",
      "   Transcript is deprecated, resolved gene is ENSG00000176155\n",
      "   Current canonical transcript is ENST00000694881\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000618014\n",
      "   Transcript is deprecated, resolved gene is ENSG00000145386\n",
      "   Current canonical transcript is ENST00000274026\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000401743\n",
      "   Transcript is deprecated, resolved gene is ENSG00000170458\n",
      "   Current canonical transcript is ENST00000302014\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000453495\n",
      "   Transcript is deprecated, resolved gene is ENSG00000099814\n",
      "   Current canonical transcript is ENST00000414716\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615636\n",
      "   Transcript is deprecated, resolved gene is ENSG00000134255\n",
      "   Current canonical transcript is ENST00000357172\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000260595\n",
      "   Transcript is deprecated, resolved gene is ENSG00000138028\n",
      "   Current canonical transcript is ENST00000402394\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000619537\n",
      "   Current canonical transcript is ENST00000619537\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000620528\n",
      "   Current canonical transcript is ENST00000620528\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000609024\n",
      "   Transcript is deprecated, resolved gene is ENSG00000254505\n",
      "   Current canonical transcript is ENST00000347519\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000619275\n",
      "   Transcript is deprecated, resolved gene is ENSG00000140839\n",
      "   Current canonical transcript is ENST00000682950\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000248228\n",
      "   Transcript is deprecated, resolved gene is ENSG00000104938\n",
      "   Current canonical transcript is ENST00000327325\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620463\n",
      "   Transcript is deprecated, resolved gene is ENSG00000164692\n",
      "   Current canonical transcript is ENST00000297268\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000545689\n",
      "   Transcript is deprecated, resolved gene is ENSG00000197565\n",
      "   Current canonical transcript is ENST00000334504\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000265379\n",
      "   Transcript is deprecated, resolved gene is ENSG00000172752\n",
      "   Current canonical transcript is ENST00000373157\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000264193\n",
      "   Transcript is deprecated, resolved gene is ENSG00000080819\n",
      "   Current canonical transcript is ENST00000647941\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000215939\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100122\n",
      "   Current canonical transcript is ENST00000647684\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000262825\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100368\n",
      "   Current canonical transcript is ENST00000403662\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000315684\n",
      "   Transcript is deprecated, resolved gene is ENSG00000178971\n",
      "   Current canonical transcript is ENST00000651323\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000372621\n",
      "   Transcript is deprecated, resolved gene is ENSG00000171793\n",
      "   Current canonical transcript is ENST00000650070\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611509\n",
      "   Transcript is deprecated, resolved gene is ENSG00000112514\n",
      "   Current canonical transcript is ENST00000488034\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000409414\n",
      "   Transcript is deprecated, resolved gene is ENSG00000197191\n",
      "   Current canonical transcript is ENST00000650725\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000545947\n",
      "   Transcript is deprecated, resolved gene is ENSG00000173402\n",
      "   Current canonical transcript is ENST00000308775\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000361951\n",
      "   Transcript is deprecated, resolved gene is ENSG00000117593\n",
      "   Current canonical transcript is ENST00000649689\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000396817\n",
      "   Current canonical transcript is ENST00000378278\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000539042\n",
      "   Transcript is deprecated, resolved gene is ENSG00000153904\n",
      "   Current canonical transcript is ENST00000284031\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000507955\n",
      "   Current canonical transcript is ENST00000330503\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000397239\n",
      "   Transcript is deprecated, resolved gene is ENSG00000124795\n",
      "   Current canonical transcript is ENST00000652689\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000409775\n",
      "   Transcript is deprecated, resolved gene is ENSG00000105928\n",
      "   Current canonical transcript is ENST00000645220\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000266991\n",
      "   Transcript is deprecated, resolved gene is ENSG00000139549\n",
      "   Current canonical transcript is ENST00000649637\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000398557\n",
      "   Transcript is deprecated, resolved gene is ENSG00000131504\n",
      "   Current canonical transcript is ENST00000389054\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000585328\n",
      "   Transcript is deprecated, resolved gene is ENSG00000187775\n",
      "   Current canonical transcript is ENST00000389840\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000342382\n",
      "   Transcript is deprecated, resolved gene is ENSG00000168724\n",
      "   Current canonical transcript is ENST00000648817\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000256935\n",
      "   Transcript is deprecated, resolved gene is ENSG00000134516\n",
      "   Current canonical transcript is ENST00000520908\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000339659\n",
      "   Transcript is deprecated, resolved gene is ENSG00000157540\n",
      "   Current canonical transcript is ENST00000647188\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000618169\n",
      "   Transcript is deprecated, resolved gene is ENSG00000129518\n",
      "   Current canonical transcript is ENST00000250454\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000563459\n",
      "   Transcript is deprecated, resolved gene is ENSG00000103356\n",
      "   Current canonical transcript is ENST00000449606\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000334286\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136160\n",
      "   Current canonical transcript is ENST00000646607\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000298049\n",
      "   Current canonical transcript is ENST00000217182\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000245323\n",
      "   Transcript is deprecated, resolved gene is ENSG00000125266\n",
      "   Current canonical transcript is ENST00000646441\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000269349\n",
      "   Transcript is deprecated, resolved gene is ENSG00000141543\n",
      "   Current canonical transcript is ENST00000649764\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000621148\n",
      "   Transcript is deprecated, resolved gene is ENSG00000242372\n",
      "   Current canonical transcript is ENST00000374450\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615377\n",
      "   Transcript is deprecated, resolved gene is ENSG00000102034\n",
      "   Current canonical transcript is ENST00000308167\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000613079\n",
      "   Transcript is deprecated, resolved gene is ENSG00000166897\n",
      "   Current canonical transcript is ENST00000402918\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000251101\n",
      "   Transcript is deprecated, resolved gene is ENSG00000132199\n",
      "   Current canonical transcript is ENST00000647584\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000360971\n",
      "   Transcript is deprecated, resolved gene is ENSG00000197594\n",
      "   Current canonical transcript is ENST00000647893\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000472938\n",
      "   Transcript is deprecated, resolved gene is ENSG00000188833\n",
      "   Current canonical transcript is ENST00000371506\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389562\n",
      "   Transcript is deprecated, resolved gene is ENSG00000183495\n",
      "   Current canonical transcript is ENST00000389561\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000621003\n",
      "   Transcript is deprecated, resolved gene is ENSG00000129595\n",
      "   Current canonical transcript is ENST00000261486\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000613142\n",
      "   Transcript is deprecated, resolved gene is ENSG00000138018\n",
      "   Current canonical transcript is ENST00000260585\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000612797\n",
      "   Transcript is deprecated, resolved gene is ENSG00000187672\n",
      "   Current canonical transcript is ENST00000288221\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000377828\n",
      "   Transcript is deprecated, resolved gene is ENSG00000187017\n",
      "   Current canonical transcript is ENST00000645284\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000556177\n",
      "   Current canonical transcript is ENST00000644823\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620555\n",
      "   Transcript is deprecated, resolved gene is ENSG00000164334\n",
      "   Current canonical transcript is ENST00000695508\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000396472\n",
      "   Transcript is deprecated, resolved gene is ENSG00000151327\n",
      "   Current canonical transcript is ENST00000280987\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622754\n",
      "   Transcript is deprecated, resolved gene is ENSG00000143409\n",
      "   Current canonical transcript is ENST00000683666\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368695\n",
      "   Transcript is deprecated, resolved gene is ENSG00000203780\n",
      "   Current canonical transcript is ENST00000368693\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000540185\n",
      "   Transcript is deprecated, resolved gene is ENSG00000164896\n",
      "   Current canonical transcript is ENST00000297532\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000427718\n",
      "   Transcript is deprecated, resolved gene is ENSG00000135108\n",
      "   Current canonical transcript is ENST00000622495\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000369384\n",
      "   Current canonical transcript is ENST00000453173\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000613418\n",
      "   Transcript is deprecated, resolved gene is ENSG00000162747\n",
      "   Current canonical transcript is ENST00000650385\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000302053\n",
      "   Transcript is deprecated, resolved gene is ENSG00000171560\n",
      "   Current canonical transcript is ENST00000403106\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000443566\n",
      "   Transcript is deprecated, resolved gene is ENSG00000187239\n",
      "   Current canonical transcript is ENST00000446176\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000380874\n",
      "   Transcript is deprecated, resolved gene is ENSG00000054598\n",
      "   Current canonical transcript is ENST00000645831\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000330315\n",
      "   Transcript is deprecated, resolved gene is ENSG00000183770\n",
      "   Current canonical transcript is ENST00000648323\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000422223\n",
      "   Transcript is deprecated, resolved gene is ENSG00000164946\n",
      "   Current canonical transcript is ENST00000380880\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000310160\n",
      "   Transcript is deprecated, resolved gene is ENSG00000174951\n",
      "   Current canonical transcript is ENST00000645652\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000505428\n",
      "   Transcript is deprecated, resolved gene is ENSG00000082074\n",
      "   Current canonical transcript is ENST00000512982\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000621232\n",
      "   Transcript is deprecated, resolved gene is ENSG00000160211\n",
      "   Current canonical transcript is ENST00000393562\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000628124\n",
      "   Transcript is deprecated, resolved gene is ENSG00000166206\n",
      "   Current canonical transcript is ENST00000311550\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000568107\n",
      "   Transcript is deprecated, resolved gene is ENSG00000261609\n",
      "   Current canonical transcript is ENST00000648994\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622544\n",
      "   Transcript is deprecated, resolved gene is ENSG00000131979\n",
      "   Current canonical transcript is ENST00000491895\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000229416\n",
      "   Transcript is deprecated, resolved gene is ENSG00000001084\n",
      "   Current canonical transcript is ENST00000650454\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000322315\n",
      "   Transcript is deprecated, resolved gene is ENSG00000179855\n",
      "   Current canonical transcript is ENST00000644452\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620807\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100938\n",
      "   Current canonical transcript is ENST00000399440\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611961\n",
      "   Transcript is deprecated, resolved gene is ENSG00000185245\n",
      "   Current canonical transcript is ENST00000329125\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000409774\n",
      "   Transcript is deprecated, resolved gene is ENSG00000152133\n",
      "   Current canonical transcript is ENST00000674370\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368232\n",
      "   Current canonical transcript is ENST00000497287\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000620890\n",
      "   Transcript is deprecated, resolved gene is ENSG00000233276\n",
      "   Current canonical transcript is ENST00000419783\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000612794\n",
      "   Transcript is deprecated, resolved gene is ENSG00000176153\n",
      "   Current canonical transcript is ENST00000389614\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622181\n",
      "   Transcript is deprecated, resolved gene is ENSG00000211445\n",
      "   Current canonical transcript is ENST00000388825\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000251808\n",
      "   Transcript is deprecated, resolved gene is ENSG00000083307\n",
      "   Current canonical transcript is ENST00000646743\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000216951\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100983\n",
      "   Current canonical transcript is ENST00000651619\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000380299\n",
      "   Transcript is deprecated, resolved gene is ENSG00000223609\n",
      "   Current canonical transcript is ENST00000650601\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000557538\n",
      "   Current canonical transcript is ENST00000337138\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000613394\n",
      "   Transcript is deprecated, resolved gene is ENSG00000107521\n",
      "   Current canonical transcript is ENST00000361490\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000256216\n",
      "   Transcript is deprecated, resolved gene is ENSG00000133835\n",
      "   Current canonical transcript is ENST00000510025\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000372425\n",
      "   Transcript is deprecated, resolved gene is ENSG00000178922\n",
      "   Current canonical transcript is ENST00000372430\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000263642\n",
      "   Transcript is deprecated, resolved gene is ENSG00000115267\n",
      "   Current canonical transcript is ENST00000649979\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000558762\n",
      "   Transcript is deprecated, resolved gene is ENSG00000140443\n",
      "   Current canonical transcript is ENST00000650285\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000389983\n",
      "   Current canonical transcript is ENST00000695950\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000412504\n",
      "   Transcript is deprecated, resolved gene is ENSG00000196083\n",
      "   Current canonical transcript is ENST00000447382\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000401393\n",
      "   Transcript is deprecated, resolved gene is ENSG00000128908\n",
      "   Current canonical transcript is ENST00000648947\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000361976\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198825\n",
      "   Current canonical transcript is ENST00000650623\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000396435\n",
      "   Transcript is deprecated, resolved gene is ENSG00000124313\n",
      "   Current canonical transcript is ENST00000642864\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000379389\n",
      "   Transcript is deprecated, resolved gene is ENSG00000187608\n",
      "   Current canonical transcript is ENST00000649529\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000287497\n",
      "   Transcript is deprecated, resolved gene is ENSG00000169896\n",
      "   Current canonical transcript is ENST00000544665\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622834\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198542\n",
      "   Current canonical transcript is ENST00000376180\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000378565\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136156\n",
      "   Current canonical transcript is ENST00000647800\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000543780\n",
      "   Transcript is deprecated, resolved gene is ENSG00000132465\n",
      "   Current canonical transcript is ENST00000254801\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368089\n",
      "   Transcript is deprecated, resolved gene is ENSG00000177807\n",
      "   Current canonical transcript is ENST00000644903\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000262888\n",
      "   Transcript is deprecated, resolved gene is ENSG00000104783\n",
      "   Current canonical transcript is ENST00000648319\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000406396\n",
      "   Transcript is deprecated, resolved gene is ENSG00000119537\n",
      "   Current canonical transcript is ENST00000645214\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000355193\n",
      "   Transcript is deprecated, resolved gene is ENSG00000055609\n",
      "   Current canonical transcript is ENST00000262189\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000265023\n",
      "   Transcript is deprecated, resolved gene is ENSG00000113889\n",
      "   Current canonical transcript is ENST00000644859\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000614544\n",
      "   Transcript is deprecated, resolved gene is ENSG00000134248\n",
      "   Current canonical transcript is ENST00000602318\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000612396\n",
      "   Transcript is deprecated, resolved gene is ENSG00000107929\n",
      "   Current canonical transcript is ENST00000316157\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000415258\n",
      "   Transcript is deprecated, resolved gene is ENSG00000011376\n",
      "   Current canonical transcript is ENST00000645846\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000592991\n",
      "   Current canonical transcript is ENST00000359446\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000541779\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198121\n",
      "   Current canonical transcript is ENST00000683809\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000311322\n",
      "   Transcript is deprecated, resolved gene is ENSG00000175445\n",
      "   Current canonical transcript is ENST00000650287\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000263816\n",
      "   Transcript is deprecated, resolved gene is ENSG00000081479\n",
      "   Current canonical transcript is ENST00000649046\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000500728\n",
      "   Transcript is deprecated, resolved gene is ENSG00000163956\n",
      "   Current canonical transcript is ENST00000650182\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000418342\n",
      "   Transcript is deprecated, resolved gene is ENSG00000169641\n",
      "   Current canonical transcript is ENST00000302291\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000395528\n",
      "   Transcript is deprecated, resolved gene is ENSG00000172901\n",
      "   Current canonical transcript is ENST00000357872\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000215739\n",
      "   Transcript is deprecated, resolved gene is ENSG00000099949\n",
      "   Current canonical transcript is ENST00000646124\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000348428\n",
      "   Transcript is deprecated, resolved gene is ENSG00000172175\n",
      "   Current canonical transcript is ENST00000649217\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000226578\n",
      "   Transcript is deprecated, resolved gene is ENSG00000109323\n",
      "   Current canonical transcript is ENST00000647097\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000493901\n",
      "   Current canonical transcript is ENST00000357582\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000523944\n",
      "   Transcript is deprecated, resolved gene is ENSG00000104738\n",
      "   Current canonical transcript is ENST00000649973\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000629399\n",
      "   Transcript is deprecated, resolved gene is ENSG00000112159\n",
      "   Current canonical transcript is ENST00000369393\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000599213\n",
      "   Transcript is deprecated, resolved gene is ENSG00000063322\n",
      "   Current canonical transcript is ENST00000315588\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000434178\n",
      "   Transcript is deprecated, resolved gene is ENSG00000074416\n",
      "   Current canonical transcript is ENST00000265052\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000537344\n",
      "   Transcript is deprecated, resolved gene is ENSG00000133816\n",
      "   Current canonical transcript is ENST00000683283\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000396617\n",
      "   Transcript is deprecated, resolved gene is ENSG00000196588\n",
      "   Current canonical transcript is ENST00000355630\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000591601\n",
      "   Transcript is deprecated, resolved gene is ENSG00000099875\n",
      "   Current canonical transcript is ENST00000250896\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000281317\n",
      "   Transcript is deprecated, resolved gene is ENSG00000151611\n",
      "   Current canonical transcript is ENST00000649156\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000233616\n",
      "   Transcript is deprecated, resolved gene is ENSG00000115275\n",
      "   Current canonical transcript is ENST00000448666\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000380172\n",
      "   Transcript is deprecated, resolved gene is ENSG00000099810\n",
      "   Current canonical transcript is ENST00000644715\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000216605\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100714\n",
      "   Current canonical transcript is ENST00000652337\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000625889\n",
      "   Transcript is deprecated, resolved gene is ENSG00000110921\n",
      "   Current canonical transcript is ENST00000228510\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000359786\n",
      "   Transcript is deprecated, resolved gene is ENSG00000197879\n",
      "   Current canonical transcript is ENST00000648651\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000265944\n",
      "   Transcript is deprecated, resolved gene is ENSG00000095777\n",
      "   Current canonical transcript is ENST00000642920\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000541942\n",
      "   Transcript is deprecated, resolved gene is ENSG00000171428\n",
      "   Current canonical transcript is ENST00000307719\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368063\n",
      "   Current canonical transcript is ENST00000294785\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000331837\n",
      "   Transcript is deprecated, resolved gene is ENSG00000182636\n",
      "   Current canonical transcript is ENST00000649030\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000378062\n",
      "   Transcript is deprecated, resolved gene is ENSG00000124479\n",
      "   Current canonical transcript is ENST00000642620\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000411638\n",
      "   Transcript is deprecated, resolved gene is ENSG00000185049\n",
      "   Current canonical transcript is ENST00000382882\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000627621\n",
      "   Current canonical transcript is ENST00000349945\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000371527\n",
      "   Transcript is deprecated, resolved gene is ENSG00000186416\n",
      "   Current canonical transcript is ENST00000688521\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622383\n",
      "   Transcript is deprecated, resolved gene is ENSG00000119919\n",
      "   Current canonical transcript is ENST00000344586\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000436936\n",
      "   Transcript is deprecated, resolved gene is ENSG00000140853\n",
      "   Current canonical transcript is ENST00000688547\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000344089\n",
      "   Transcript is deprecated, resolved gene is ENSG00000089250\n",
      "   Current canonical transcript is ENST00000317775\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000277541\n",
      "   Transcript is deprecated, resolved gene is ENSG00000148400\n",
      "   Current canonical transcript is ENST00000651671\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620134\n",
      "   Transcript is deprecated, resolved gene is ENSG00000103148\n",
      "   Current canonical transcript is ENST00000611875\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000405342\n",
      "   Transcript is deprecated, resolved gene is ENSG00000122643\n",
      "   Current canonical transcript is ENST00000610140\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000376214\n",
      "   Transcript is deprecated, resolved gene is ENSG00000148053\n",
      "   Current canonical transcript is ENST00000277120\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000379806\n",
      "   Transcript is deprecated, resolved gene is ENSG00000131828\n",
      "   Current canonical transcript is ENST00000422285\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000542749\n",
      "   Transcript is deprecated, resolved gene is ENSG00000090975\n",
      "   Current canonical transcript is ENST00000320201\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000370184\n",
      "   Current canonical transcript is ENST00000370185\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000372389\n",
      "   Transcript is deprecated, resolved gene is ENSG00000171453\n",
      "   Current canonical transcript is ENST00000642195\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000614426\n",
      "   Current canonical transcript is ENST00000331373\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000404875\n",
      "   Current canonical transcript is ENST00000402686\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000303531\n",
      "   Transcript is deprecated, resolved gene is ENSG00000166501\n",
      "   Current canonical transcript is ENST00000643927\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615510\n",
      "   Transcript is deprecated, resolved gene is ENSG00000172179\n",
      "   Current canonical transcript is ENST00000306482\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000357304\n",
      "   Transcript is deprecated, resolved gene is ENSG00000130723\n",
      "   No current transcript found!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Transcript is deprecated, resolved gene is ENSG00000275896\n",
      "   Current canonical transcript is ENST00000539842\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000606149\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100519\n",
      "   Current canonical transcript is ENST00000445930\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000620216\n",
      "   Transcript is deprecated, resolved gene is ENSG00000099341\n",
      "   Current canonical transcript is ENST00000215071\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000619580\n",
      "   Transcript is deprecated, resolved gene is ENSG00000159335\n",
      "   Current canonical transcript is ENST00000309083\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000306726\n",
      "   Transcript is deprecated, resolved gene is ENSG00000169410\n",
      "   Current canonical transcript is ENST00000618819\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000403432\n",
      "   Transcript is deprecated, resolved gene is ENSG00000055917\n",
      "   Current canonical transcript is ENST00000361078\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000527788\n",
      "   Current canonical transcript is ENST00000650167\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000329286\n",
      "   Transcript is deprecated, resolved gene is ENSG00000206418\n",
      "   Current canonical transcript is ENST00000649141\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000610732\n",
      "   Transcript is deprecated, resolved gene is ENSG00000117602\n",
      "   Current canonical transcript is ENST00000374395\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000343115\n",
      "   Transcript is deprecated, resolved gene is ENSG00000137710\n",
      "   Current canonical transcript is ENST00000645495\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000625761\n",
      "   Transcript is deprecated, resolved gene is ENSG00000104856\n",
      "   Current canonical transcript is ENST00000221452\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611665\n",
      "   Transcript is deprecated, resolved gene is ENSG00000076043\n",
      "   Current canonical transcript is ENST00000265881\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000617334\n",
      "   Transcript is deprecated, resolved gene is ENSG00000132554\n",
      "   Current canonical transcript is ENST00000360863\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000617441\n",
      "   Transcript is deprecated, resolved gene is ENSG00000168421\n",
      "   Current canonical transcript is ENST00000381799\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000556710\n",
      "   Current canonical transcript is ENST00000485079\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000318842\n",
      "   Transcript is deprecated, resolved gene is ENSG00000156313\n",
      "   Current canonical transcript is ENST00000645032\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000424325\n",
      "   Transcript is deprecated, resolved gene is ENSG00000147403\n",
      "   Current canonical transcript is ENST00000369817\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000464167\n",
      "   Transcript is deprecated, resolved gene is ENSG00000182899\n",
      "   Current canonical transcript is ENST00000647248\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368567\n",
      "   Transcript is deprecated, resolved gene is ENSG00000177954\n",
      "   Current canonical transcript is ENST00000651669\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000534062\n",
      "   Transcript is deprecated, resolved gene is ENSG00000254656\n",
      "   Current canonical transcript is ENST00000649591\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000590320\n",
      "   Transcript is deprecated, resolved gene is ENSG00000267534\n",
      "   Current canonical transcript is ENST00000646641\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000617721\n",
      "   Transcript is deprecated, resolved gene is ENSG00000180739\n",
      "   Current canonical transcript is ENST00000333430\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611159\n",
      "   Transcript is deprecated, resolved gene is ENSG00000179134\n",
      "   Current canonical transcript is ENST00000610417\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000621965\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136546\n",
      "   Current canonical transcript is ENST00000643258\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000377475\n",
      "   Transcript is deprecated, resolved gene is ENSG00000101310\n",
      "   Current canonical transcript is ENST00000650089\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000541726\n",
      "   Transcript is deprecated, resolved gene is ENSG00000113811\n",
      "   Current canonical transcript is ENST00000495461\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611680\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198832\n",
      "   Current canonical transcript is ENST00000400299\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611222\n",
      "   Transcript is deprecated, resolved gene is ENSG00000073169\n",
      "   Current canonical transcript is ENST00000380903\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615547\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198843\n",
      "   Current canonical transcript is ENST00000471696\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368286\n",
      "   Transcript is deprecated, resolved gene is ENSG00000196189\n",
      "   Current canonical transcript is ENST00000368285\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000616787\n",
      "   Transcript is deprecated, resolved gene is ENSG00000183291\n",
      "   Current canonical transcript is ENST00000331835\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000612212\n",
      "   Transcript is deprecated, resolved gene is ENSG00000178980\n",
      "   Current canonical transcript is ENST00000601048\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000367104\n",
      "   Transcript is deprecated, resolved gene is ENSG00000122335\n",
      "   Current canonical transcript is ENST00000647468\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000616363\n",
      "   Transcript is deprecated, resolved gene is ENSG00000114204\n",
      "   Current canonical transcript is ENST00000264677\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000330479\n",
      "   Transcript is deprecated, resolved gene is ENSG00000183955\n",
      "   Current canonical transcript is ENST00000402868\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000265735\n",
      "   Transcript is deprecated, resolved gene is ENSG00000127990\n",
      "   Current canonical transcript is ENST00000648936\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000357436\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100092\n",
      "   Current canonical transcript is ENST00000649765\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000534261\n",
      "   Current canonical transcript is ENST00000534261\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000558405\n",
      "   Transcript is deprecated, resolved gene is ENSG00000074803\n",
      "   Current canonical transcript is ENST00000380993\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000298472\n",
      "   Transcript is deprecated, resolved gene is ENSG00000165646\n",
      "   Current canonical transcript is ENST00000644641\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000397260\n",
      "   Current canonical transcript is ENST00000354772\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000273158\n",
      "   Transcript is deprecated, resolved gene is ENSG00000144659\n",
      "   Current canonical transcript is ENST00000650617\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000427851\n",
      "   Transcript is deprecated, resolved gene is ENSG00000112759\n",
      "   Current canonical transcript is ENST00000371755\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000392845\n",
      "   Transcript is deprecated, resolved gene is ENSG00000169359\n",
      "   Current canonical transcript is ENST00000643144\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000424970\n",
      "   Current canonical transcript is ENST00000356736\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000539553\n",
      "   Transcript is deprecated, resolved gene is ENSG00000088836\n",
      "   Current canonical transcript is ENST00000642402\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000400286\n",
      "   Transcript is deprecated, resolved gene is ENSG00000184564\n",
      "   Current canonical transcript is ENST00000647374\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000614072\n",
      "   Transcript is deprecated, resolved gene is ENSG00000171201\n",
      "   Current canonical transcript is ENST00000304915\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000602728\n",
      "   Transcript is deprecated, resolved gene is ENSG00000114902\n",
      "   Current canonical transcript is ENST00000619898\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000556626\n",
      "   Transcript is deprecated, resolved gene is ENSG00000070182\n",
      "   Current canonical transcript is ENST00000644917\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000342756\n",
      "   Transcript is deprecated, resolved gene is ENSG00000174780\n",
      "   Current canonical transcript is ENST00000642900\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000586243\n",
      "   Transcript is deprecated, resolved gene is ENSG00000118046\n",
      "   Current canonical transcript is ENST00000326873\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000267540\n",
      "   Transcript is deprecated, resolved gene is ENSG00000140022\n",
      "   Current canonical transcript is ENST00000614646\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000378654\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136143\n",
      "   Current canonical transcript is ENST00000646932\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000538427\n",
      "   Transcript is deprecated, resolved gene is ENSG00000055208\n",
      "   Current canonical transcript is ENST00000637181\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615887\n",
      "   Transcript is deprecated, resolved gene is ENSG00000166863\n",
      "   Current canonical transcript is ENST00000458521\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000432496\n",
      "   Transcript is deprecated, resolved gene is ENSG00000145348\n",
      "   Current canonical transcript is ENST00000394708\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000616883\n",
      "   Transcript is deprecated, resolved gene is ENSG00000112592\n",
      "   Current canonical transcript is ENST00000392092\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000609199\n",
      "   Current canonical transcript is ENST00000613537\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000611869\n",
      "   Transcript is deprecated, resolved gene is ENSG00000071564\n",
      "   Current canonical transcript is ENST00000262965\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000536810\n",
      "   Transcript is deprecated, resolved gene is ENSG00000148737\n",
      "   Current canonical transcript is ENST00000355995\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000535286\n",
      "   Transcript is deprecated, resolved gene is ENSG00000083544\n",
      "   Current canonical transcript is ENST00000377881\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000613071\n",
      "   Transcript is deprecated, resolved gene is ENSG00000180190\n",
      "   Current canonical transcript is ENST00000324079\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000615733\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136478\n",
      "   Current canonical transcript is ENST00000584379\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000558083\n",
      "   Current canonical transcript is ENST00000696102\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000392129\n",
      "   Transcript is deprecated, resolved gene is ENSG00000143183\n",
      "   Current canonical transcript is ENST00000367881\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000610315\n",
      "   Transcript is deprecated, resolved gene is ENSG00000261115\n",
      "   Current canonical transcript is ENST00000565468\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000617947\n",
      "   Transcript is deprecated, resolved gene is ENSG00000130598\n",
      "   Current canonical transcript is ENST00000381911\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000588847\n",
      "   Current canonical transcript is ENST00000696270\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000392451\n",
      "   Transcript is deprecated, resolved gene is ENSG00000163462\n",
      "   Current canonical transcript is ENST00000334634\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000406386\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100106\n",
      "   Current canonical transcript is ENST00000644935\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000290846\n",
      "   Transcript is deprecated, resolved gene is ENSG00000100416\n",
      "   Current canonical transcript is ENST00000645190\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000454289\n",
      "   Transcript is deprecated, resolved gene is ENSG00000123297\n",
      "   Current canonical transcript is ENST00000652027\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000622375\n",
      "   Transcript is deprecated, resolved gene is ENSG00000128881\n",
      "   Current canonical transcript is ENST00000267890\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000368563\n",
      "   Transcript is deprecated, resolved gene is ENSG00000130640\n",
      "   Current canonical transcript is ENST00000252936\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000429002\n",
      "   Transcript is deprecated, resolved gene is ENSG00000198431\n",
      "   Current canonical transcript is ENST00000525566\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000610664\n",
      "   Current canonical transcript is ENST00000610664\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "processing ENST00000367924\n",
      "   Transcript is deprecated, resolved gene is ENSG00000117143\n",
      "   Current canonical transcript is ENST00000367925\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000367274\n",
      "   Transcript is deprecated, resolved gene is ENSG00000077152\n",
      "   Current canonical transcript is ENST00000646651\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000614096\n",
      "   Transcript is deprecated, resolved gene is ENSG00000114062\n",
      "   Current canonical transcript is ENST00000648336\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000616789\n",
      "   Transcript is deprecated, resolved gene is ENSG00000105668\n",
      "   Current canonical transcript is ENST00000222275\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000261517\n",
      "   Transcript is deprecated, resolved gene is ENSG00000129003\n",
      "   Current canonical transcript is ENST00000644861\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000369130\n",
      "   Transcript is deprecated, resolved gene is ENSG00000136631\n",
      "   Current canonical transcript is ENST00000644510\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000618527\n",
      "   Transcript is deprecated, resolved gene is ENSG00000062650\n",
      "   Current canonical transcript is ENST00000298767\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000434753\n",
      "   Transcript is deprecated, resolved gene is ENSG00000101966\n",
      "   Current canonical transcript is ENST00000371199\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000400602\n",
      "   Transcript is deprecated, resolved gene is ENSG00000132953\n",
      "   Current canonical transcript is ENST00000255305\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000373452\n",
      "   Transcript is deprecated, resolved gene is ENSG00000177125\n",
      "   Current canonical transcript is ENST00000319119\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000545588\n",
      "   Transcript is deprecated, resolved gene is ENSG00000058673\n",
      "   Current canonical transcript is ENST00000367210\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000309776\n",
      "   Transcript is deprecated, resolved gene is ENSG00000015171\n",
      "   Current canonical transcript is ENST00000381604\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000534834\n",
      "   Current canonical transcript is ENST00000328654\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000309495\n",
      "   Transcript is deprecated, resolved gene is ENSG00000175105\n",
      "   Current canonical transcript is ENST00000636215\n",
      "   FA and BED files found. Updating dataframe with current information\n",
      "processing ENST00000543588\n",
      "   Transcript is deprecated, resolved gene is ENSG00000166529\n",
      "   Current canonical transcript is ENST00000292450\n",
      "   FA and BED files found. Updating dataframe with current information\n"
     ]
    }
   ],
   "source": [
    "# loop over all transcript IDs without a transcript file associated with it\n",
    "\n",
    "for tid in df2.loc[ df2['bed_files'] == 0, 'EnsemblTranscriptID' ]: #.head(2):\n",
    "    print('processing', tid)\n",
    "    # Ensembl URL for resolving the given transcript ID\n",
    "    url = 'https://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=' + tid\n",
    "    # retrieve the document\n",
    "    r = requests.get(url)\n",
    "    # parse the document\n",
    "    soup = BeautifulSoup(r.content, 'html.parser')\n",
    "    # check for specific strings in the page\n",
    "    if re.search(r'This transcript is not in the current gene set', soup.get_text()):\n",
    "        table_content = soup.tbody;\n",
    "        # loop over all listed genes\n",
    "        for i in table_content.find_all_next('td'):\n",
    "            # search for columns with a gene link\n",
    "            if i.find('a') and re.search(r'.*Gene.*ENSG\\d+', i.a.attrs['href']):\n",
    "                # extract the gene ID\n",
    "                href = i.a.attrs['href']\n",
    "                gene = re.sub(r'.*(ENSG0\\d+)', r'\\1', href)\n",
    "                print('   Transcript is deprecated, resolved gene is', gene)\n",
    "\n",
    "                # Ensembl URL to resolve a gene ID\n",
    "                url = 'https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=' + gene\n",
    "                r = requests.get(url)\n",
    "                new_tid = find_new_transcript(r.content)\n",
    "                if check_files_and_update_df(tid, new_tid): break\n",
    "    \n",
    "    elif re.search(r'Show transcript table', soup.get_text()):\n",
    "        # the transcript is not the current canonical version and needs updating\n",
    "        new_tid = find_new_transcript(r.content)\n",
    "        check_files_and_update_df(tid, new_tid)\n",
    "    else:\n",
    "        print('   Some other error occured for this transcript')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "834fbc22-ba69-4063-a639-cd7023e68465",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if there are still entries with unresolved transcript files\n",
    "# spoiler, there are 10\n",
    "df2.loc[ df2['bed_files'] != 1, 'EnsemblTranscriptID' ].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a8c2034c-e9a1-414a-a3aa-4379e25515d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "      <th>bed_files</th>\n",
       "      <th>fa_files</th>\n",
       "      <th>bed</th>\n",
       "      <th>fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>ENST00000366779</td>\n",
       "      <td>5.703</td>\n",
       "      <td>5.190</td>\n",
       "      <td>5.879</td>\n",
       "      <td>5.753</td>\n",
       "      <td>5.569</td>\n",
       "      <td>5.699</td>\n",
       "      <td>5.663</td>\n",
       "      <td>5.875</td>\n",
       "      <td>5.788</td>\n",
       "      <td>...</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.4896</td>\n",
       "      <td>5.586</td>\n",
       "      <td>6.075</td>\n",
       "      <td>5.588</td>\n",
       "      <td>5.0744</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>ENST00000382387</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>ENST00000624406</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.201</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.864</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>ENST00000619537</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941</th>\n",
       "      <td>ENST00000620528</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.221</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>ENST00000369384</td>\n",
       "      <td>4.442</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.537</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.2395</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4224</th>\n",
       "      <td>ENST00000368232</td>\n",
       "      <td>4.089</td>\n",
       "      <td>3.836</td>\n",
       "      <td>3.029</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.773</td>\n",
       "      <td>3.914</td>\n",
       "      <td>4.630</td>\n",
       "      <td>4.463</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>3.664</td>\n",
       "      <td>4.2940</td>\n",
       "      <td>3.668</td>\n",
       "      <td>4.027</td>\n",
       "      <td>3.985</td>\n",
       "      <td>2.9934</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7810</th>\n",
       "      <td>ENST00000357304</td>\n",
       "      <td>4.154</td>\n",
       "      <td>3.726</td>\n",
       "      <td>3.551</td>\n",
       "      <td>4.201</td>\n",
       "      <td>3.729</td>\n",
       "      <td>3.665</td>\n",
       "      <td>3.392</td>\n",
       "      <td>4.057</td>\n",
       "      <td>3.011</td>\n",
       "      <td>...</td>\n",
       "      <td>3.515</td>\n",
       "      <td>3.7242</td>\n",
       "      <td>3.830</td>\n",
       "      <td>3.025</td>\n",
       "      <td>3.805</td>\n",
       "      <td>3.5673</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9042</th>\n",
       "      <td>ENST00000534261</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.356</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.152</td>\n",
       "      <td>...</td>\n",
       "      <td>5.351</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.5499</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10776</th>\n",
       "      <td>ENST00000610664</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>6.141</td>\n",
       "      <td>...</td>\n",
       "      <td>7.237</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EnsemblTranscriptID Adrenal_PTR Appendices_PTR Brain_PTR Colon_PTR  \\\n",
       "220       ENST00000366779       5.703          5.190     5.879     5.753   \n",
       "1377      ENST00000382387          NA             NA        NA        NA   \n",
       "1532      ENST00000624406          NA             NA     7.201        NA   \n",
       "1940      ENST00000619537          NA             NA        NA        NA   \n",
       "1941      ENST00000620528          NA             NA        NA        NA   \n",
       "3679      ENST00000369384       4.442             NA        NA        NA   \n",
       "4224      ENST00000368232       4.089          3.836     3.029        NA   \n",
       "7810      ENST00000357304       4.154          3.726     3.551     4.201   \n",
       "9042      ENST00000534261          NA          5.356        NA        NA   \n",
       "10776     ENST00000610664          NA             NA        NA        NA   \n",
       "\n",
       "      Duodenum_PTR Endometrium_PTR Esophagus_PTR Fallopiantube_PTR Fat_PTR  \\\n",
       "220          5.569           5.699         5.663             5.875   5.788   \n",
       "1377            NA              NA            NA                NA      NA   \n",
       "1532            NA              NA            NA                NA      NA   \n",
       "1940            NA              NA            NA                NA      NA   \n",
       "1941         6.221              NA            NA                NA      NA   \n",
       "3679            NA              NA            NA                NA   4.537   \n",
       "4224         3.773           3.914         4.630             4.463      NA   \n",
       "7810         3.729           3.665         3.392             4.057   3.011   \n",
       "9042            NA              NA            NA                NA   5.152   \n",
       "10776           NA              NA            NA                NA   6.141   \n",
       "\n",
       "       ... Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR Tonsil_PTR  \\\n",
       "220    ...      5.582      5.4896      5.586       6.075      5.588   \n",
       "1377   ...         NA          NA         NA          NA         NA   \n",
       "1532   ...         NA          NA         NA       6.864         NA   \n",
       "1940   ...         NA          NA         NA          NA         NA   \n",
       "1941   ...         NA          NA         NA          NA         NA   \n",
       "3679   ...         NA          NA         NA          NA         NA   \n",
       "4224   ...      3.664      4.2940      3.668       4.027      3.985   \n",
       "7810   ...      3.515      3.7242      3.830       3.025      3.805   \n",
       "9042   ...      5.351          NA         NA          NA         NA   \n",
       "10776  ...      7.237          NA         NA          NA         NA   \n",
       "\n",
       "      Urinarybladder_PTR bed_files fa_files bed fa  \n",
       "220               5.0744         0        0         \n",
       "1377                  NA         0        0         \n",
       "1532                  NA         0        0         \n",
       "1940                  NA         0        0         \n",
       "1941                  NA         0        0         \n",
       "3679              5.2395         0        0         \n",
       "4224              2.9934         0        0         \n",
       "7810              3.5673         0        0         \n",
       "9042              5.5499         0        0         \n",
       "10776                 NA         0        0         \n",
       "\n",
       "[10 rows x 34 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show the 10 IDs with missing transcript files\n",
    "df2.loc[ df2['bed_files'] == 0 ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f006f00f-d749-420a-9624-7ec1c65b5d66",
   "metadata": {},
   "source": [
    "Even though these transcript IDs will have resolved in 2019 when the paper using this data was published the current database does not resolve these IDs any more (due to more up to date research results).\n",
    "\n",
    "Being brave those entries will be deleted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62f12fc-f489-4b9a-a47d-5abfec3006cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove the 11 missing entries\n",
    "df2.drop(df2[df2['bed_files'] == 0].index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "456ac807-72f6-49e4-a591-f3f02bcd5711",
   "metadata": {},
   "outputs": [],
   "source": [
    "# verify the entries are gone from the dataframe\n",
    "df2.loc[ df2['bed_files'] == 0 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0287abdd-0144-4268-adef-c37b763e7eae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Spleen_PTR</th>\n",
       "      <th>Stomach_PTR</th>\n",
       "      <th>Testis_PTR</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "      <th>bed_files</th>\n",
       "      <th>fa_files</th>\n",
       "      <th>bed</th>\n",
       "      <th>fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.277</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.841</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>7.313</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.135</td>\n",
       "      <td>5.371</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.8143</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>6.290</td>\n",
       "      <td>6.328</td>\n",
       "      <td>5.948</td>\n",
       "      <td>5.811</td>\n",
       "      <td>6.068</td>\n",
       "      <td>5.383</td>\n",
       "      <td>5.881</td>\n",
       "      <td>6.119</td>\n",
       "      <td>6.410</td>\n",
       "      <td>...</td>\n",
       "      <td>5.136</td>\n",
       "      <td>6.5349</td>\n",
       "      <td>5.820</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.995</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.129</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>2.350</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>3.843</td>\n",
       "      <td>4.601</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.013</td>\n",
       "      <td>3.683</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.0613</td>\n",
       "      <td>4.832</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11570</th>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11571</th>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>4.461</td>\n",
       "      <td>5.013</td>\n",
       "      <td>5.047</td>\n",
       "      <td>4.566</td>\n",
       "      <td>5.184</td>\n",
       "      <td>4.826</td>\n",
       "      <td>5.102</td>\n",
       "      <td>4.670</td>\n",
       "      <td>5.756</td>\n",
       "      <td>...</td>\n",
       "      <td>4.255</td>\n",
       "      <td>4.0412</td>\n",
       "      <td>5.389</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11572</th>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.524</td>\n",
       "      <td>5.478</td>\n",
       "      <td>5.915</td>\n",
       "      <td>5.811</td>\n",
       "      <td>5.817</td>\n",
       "      <td>5.943</td>\n",
       "      <td>5.509</td>\n",
       "      <td>4.931</td>\n",
       "      <td>...</td>\n",
       "      <td>6.159</td>\n",
       "      <td>5.8846</td>\n",
       "      <td>5.582</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11573</th>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>5.112</td>\n",
       "      <td>4.918</td>\n",
       "      <td>5.139</td>\n",
       "      <td>5.190</td>\n",
       "      <td>5.442</td>\n",
       "      <td>5.602</td>\n",
       "      <td>4.715</td>\n",
       "      <td>4.956</td>\n",
       "      <td>5.033</td>\n",
       "      <td>...</td>\n",
       "      <td>5.110</td>\n",
       "      <td>5.0834</td>\n",
       "      <td>5.047</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>3.165</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.539</td>\n",
       "      <td>3.629</td>\n",
       "      <td>3.616</td>\n",
       "      <td>4.203</td>\n",
       "      <td>3.400</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.4766</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>../../GENCODE43/protein_coding/BED6__protein_c...</td>\n",
       "      <td>../../GENCODE43/protein_coding/FA_protein_codi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11575 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EnsemblTranscriptID Adrenal_PTR Appendices_PTR Brain_PTR Colon_PTR  \\\n",
       "0         ENST00000263100          NA          8.277        NA        NA   \n",
       "1         ENST00000373993          NA             NA        NA     5.135   \n",
       "2         ENST00000318602       6.290          6.328     5.948     5.811   \n",
       "3         ENST00000299698          NA             NA     3.995        NA   \n",
       "4         ENST00000401850       3.843          4.601        NA        NA   \n",
       "...                   ...         ...            ...       ...       ...   \n",
       "11570     ENST00000374888          NA             NA        NA        NA   \n",
       "11571     ENST00000294353       4.461          5.013     5.047     4.566   \n",
       "11572     ENST00000322764       5.664          5.524     5.478     5.915   \n",
       "11573     ENST00000381638       5.112          4.918     5.139     5.190   \n",
       "11574     ENST00000370801       3.165             NA        NA     3.539   \n",
       "\n",
       "      Duodenum_PTR Endometrium_PTR Esophagus_PTR Fallopiantube_PTR Fat_PTR  \\\n",
       "0               NA              NA            NA             7.841      NA   \n",
       "1            5.371              NA            NA                NA      NA   \n",
       "2            6.068           5.383         5.881             6.119   6.410   \n",
       "3               NA              NA         4.129                NA      NA   \n",
       "4               NA              NA         4.013             3.683      NA   \n",
       "...            ...             ...           ...               ...     ...   \n",
       "11570           NA              NA            NA                NA      NA   \n",
       "11571        5.184           4.826         5.102             4.670   5.756   \n",
       "11572        5.811           5.817         5.943             5.509   4.931   \n",
       "11573        5.442           5.602         4.715             4.956   5.033   \n",
       "11574        3.629           3.616         4.203             3.400      NA   \n",
       "\n",
       "       ... Spleen_PTR Stomach_PTR Testis_PTR Thyroid_PTR Tonsil_PTR  \\\n",
       "0      ...      7.313          NA         NA          NA         NA   \n",
       "1      ...         NA      5.8143         NA          NA         NA   \n",
       "2      ...      5.136      6.5349      5.820       6.060      5.675   \n",
       "3      ...         NA          NA      2.350          NA      5.249   \n",
       "4      ...         NA      4.0613      4.832          NA         NA   \n",
       "...    ...        ...         ...        ...         ...        ...   \n",
       "11570  ...         NA          NA         NA          NA      4.681   \n",
       "11571  ...      4.255      4.0412      5.389       4.250      4.439   \n",
       "11572  ...      6.159      5.8846      5.582       5.598      5.968   \n",
       "11573  ...      5.110      5.0834      5.047       5.038      5.130   \n",
       "11574  ...         NA      3.4766         NA          NA      3.008   \n",
       "\n",
       "      Urinarybladder_PTR bed_files fa_files  \\\n",
       "0                     NA         1        1   \n",
       "1                     NA         1        1   \n",
       "2                 5.8286         1        1   \n",
       "3                     NA         1        1   \n",
       "4                 4.2430         1        1   \n",
       "...                  ...       ...      ...   \n",
       "11570                 NA         1        1   \n",
       "11571             4.1460         1        1   \n",
       "11572             5.3358         1        1   \n",
       "11573             5.0619         1        1   \n",
       "11574             3.6041         1        1   \n",
       "\n",
       "                                                     bed  \\\n",
       "0      ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "1      ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "2      ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "3      ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "4      ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "...                                                  ...   \n",
       "11570  ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "11571  ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "11572  ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "11573  ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "11574  ../../GENCODE43/protein_coding/BED6__protein_c...   \n",
       "\n",
       "                                                      fa  \n",
       "0      ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "1      ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "2      ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "3      ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "4      ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "...                                                  ...  \n",
       "11570  ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "11571  ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "11572  ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "11573  ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "11574  ../../GENCODE43/protein_coding/FA_protein_codi...  \n",
       "\n",
       "[11575 rows x 34 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show the rest again\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f910f9c-58c4-49a3-9466-bc734035ee89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write current pre processed table to file\n",
    "datafile = '../data/preproc_stage1.csv'\n",
    "df2.to_csv(datafile, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
