{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "822ec892-2452-4341-95c9-a1c65018f6f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from pathlib import Path\n",
    "from bs4 import BeautifulSoup\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "05ce14fc-cd68-4271-9db8-39c219232181",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw data paths\n",
    "gencode_path = '../../GENCODE43/protein_coding/'\n",
    "bed = Path(gencode_path) / 'BED6__protein_coding_strict/'\n",
    "fa = Path(gencode_path) / 'FA_protein_coding_strict_mRNA/'\n",
    "\n",
    "def find_new_transcript(content):\n",
    "    \"\"\"Method to extract a specific transcript ID from an HTML document.\n",
    "\n",
    "    Keyword Arguments:\n",
    "    req -- Python request object content\n",
    "\n",
    "    Returns:\n",
    "    A string either empty or containing the transcript ID.\n",
    "    \"\"\"\n",
    "    # parse the HTML document\n",
    "    soup = BeautifulSoup(content, 'html.parser')\n",
    "    # check if a specific table exists\n",
    "    if soup.find(id='transcripts_table'):\n",
    "        # if so extract the transcript ID\n",
    "        href = soup.find(id='transcripts_table').tbody.td.a.attrs['href']\n",
    "        transcript = re.sub(r'.*(ENST0\\d+)', r'\\1', href)\n",
    "        print('   Current canonical transcript is', transcript)\n",
    "    else:\n",
    "        # if not return an empty string\n",
    "        transcript = ''\n",
    "        print('   No current transcript found!')\n",
    "\n",
    "    return transcript\n",
    "\n",
    "def check_files_and_update_df(old_tid, new_tid):\n",
    "    \"\"\"Cross reference the transcript ID with files in the gencode data set\n",
    "    (bad hack as it uses variables globally defined at the beginning of this notebook!)\n",
    "\n",
    "    Keyword Arguments:\n",
    "    transcript -- the transcript ID string\n",
    "    \"\"\"\n",
    "    # search and count files with a given name\n",
    "    bed_file_list = list(bed.glob(new_tid + '*.bed'))\n",
    "    bed_files = len(bed_file_list)\n",
    "    fa_file_list = list(fa.glob(new_tid + '*.fasta'))\n",
    "    fa_files = len(fa_file_list)\n",
    "\n",
    "    # check how many files were found\n",
    "    if bed_files == 1 and fa_files == 1:\n",
    "        # if it's 1 everything is perfect\n",
    "        print('   FA and BED files found. Updating dataframe with current information')\n",
    "        # update dataframe\n",
    "        # df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'bed_files' ] = bed_files\n",
    "        # df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'fa_files' ] = fa_files\n",
    "        # df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'bed'] = str(bed_file_list[0])\n",
    "        # df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'fa'] = str(fa_file_list[0])\n",
    "        # df2.loc[ df2['EnsemblTranscriptID'] == old_tid, 'EnsemblTranscriptID' ] = new_tid\n",
    "        return True\n",
    "    else:\n",
    "        # if there are many manual processing is needed\n",
    "        print('   FA and BED file count invalid. File lists', bed_file_list, fa_file_list)\n",
    "\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "39e8b8fc-0bcc-4f1f-875d-6c07174cb1a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://www.ensembl.org/Homo_sapiens/Transcript/Idhistory?t=ENST00000442201'\n",
    "r = requests.get(url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "816fd6f7-1ddd-4951-9d9e-05d4387aa0b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Transcript is deprecated, resolved gene is ENSG00000145075\n",
      "   Current canonical transcript is ENST00000471307\n",
      "   FA and BED file count invalid. File lists [] []\n",
      "   Transcript is deprecated, resolved gene is ENSG00000284862\n",
      "   Current canonical transcript is ENST00000476379\n",
      "   FA and BED files found. Updating dataframe with current information\n"
     ]
    }
   ],
   "source": [
    "soup = BeautifulSoup(r.content, 'html.parser')\n",
    "\n",
    "table_content = soup.tbody;\n",
    "\n",
    "# print(table_content)\n",
    "\n",
    "tid = 'ENST00000442201'\n",
    "\n",
    "for i in table_content.find_all_next('td'):\n",
    "    if i.find('a') and re.search(r'.*Gene.*ENSG\\d+', i.a.attrs['href']):\n",
    "        href = i.a.attrs['href']\n",
    "        gene = re.sub(r'.*(ENSG0\\d+)', r'\\1', href)\n",
    "        print('   Transcript is deprecated, resolved gene is', gene)\n",
    "\n",
    "        url = 'https://www.ensembl.org/Homo_sapiens/Gene/Idhistory?g=' + gene\n",
    "        r = requests.get(url)\n",
    "        new_tid = find_new_transcript(r.content)\n",
    "        if check_files_and_update_df(tid, new_tid): break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ace2b15f-9950-49d5-a477-599e6a962c87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2274'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tid = 'ENST00000000442'\n",
    "line = '>ENST00000000442.11:0-2274'\n",
    "\n",
    "m = re.match(r'>' + tid + r'\\.\\d+:0-(?P<trans_length>\\d+)', line)\n",
    "\n",
    "m.group('trans_length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "51c83f85-3889-4ac4-ab79-c9b42f7afe44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblTranscriptID</th>\n",
       "      <th>Adrenal_PTR</th>\n",
       "      <th>Appendices_PTR</th>\n",
       "      <th>Brain_PTR</th>\n",
       "      <th>Colon_PTR</th>\n",
       "      <th>Duodenum_PTR</th>\n",
       "      <th>Endometrium_PTR</th>\n",
       "      <th>Esophagus_PTR</th>\n",
       "      <th>Fallopiantube_PTR</th>\n",
       "      <th>Fat_PTR</th>\n",
       "      <th>...</th>\n",
       "      <th>Thyroid_PTR</th>\n",
       "      <th>Tonsil_PTR</th>\n",
       "      <th>Urinarybladder_PTR</th>\n",
       "      <th>5UTR start</th>\n",
       "      <th>5UTR stop</th>\n",
       "      <th>CDS start</th>\n",
       "      <th>CDS stop</th>\n",
       "      <th>3UTR start</th>\n",
       "      <th>3UTR stop</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>NA</td>\n",
       "      <td>8.277</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>7.841</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>1543</td>\n",
       "      <td>1543</td>\n",
       "      <td>3382</td>\n",
       "      <td>ATTGCTGCAGACGCTCACCCCAGACACTCACTGCACCGGAGTGAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000373993</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.135</td>\n",
       "      <td>5.371</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>1877</td>\n",
       "      <td>1877</td>\n",
       "      <td>2044</td>\n",
       "      <td>ATAATCAAGGAAACCTTTTCCGGGTGGGGATCTCTGAAATTACTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>6.290</td>\n",
       "      <td>6.328</td>\n",
       "      <td>5.948</td>\n",
       "      <td>5.811</td>\n",
       "      <td>6.068</td>\n",
       "      <td>5.383</td>\n",
       "      <td>5.881</td>\n",
       "      <td>6.119</td>\n",
       "      <td>6.410</td>\n",
       "      <td>...</td>\n",
       "      <td>6.060</td>\n",
       "      <td>5.675</td>\n",
       "      <td>5.8286</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>4495</td>\n",
       "      <td>4495</td>\n",
       "      <td>4610</td>\n",
       "      <td>GGGACCAGATGGATTGTAGGGAGTAGGGTACAATACAGTCTGTTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.995</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.129</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>5.249</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>4396</td>\n",
       "      <td>4396</td>\n",
       "      <td>5127</td>\n",
       "      <td>GACCCTGGAAAAATCTGTCTCACCCACAAAGATGTGGGCTCAGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>3.843</td>\n",
       "      <td>4.601</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.013</td>\n",
       "      <td>3.683</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.2430</td>\n",
       "      <td>0</td>\n",
       "      <td>490</td>\n",
       "      <td>490</td>\n",
       "      <td>1552</td>\n",
       "      <td>1552</td>\n",
       "      <td>2321</td>\n",
       "      <td>TGCACTTCTGTGCCTCAATTTCCTCATCTGTAGGGTGGGGGTGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11560</th>\n",
       "      <td>ENST00000374888</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>4.681</td>\n",
       "      <td>NA</td>\n",
       "      <td>0</td>\n",
       "      <td>38</td>\n",
       "      <td>38</td>\n",
       "      <td>2450</td>\n",
       "      <td>2450</td>\n",
       "      <td>5467</td>\n",
       "      <td>CTCGGCTCTGGTTCCAGCCGAGCCTCTCGGACGCAGAGATGGAAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11561</th>\n",
       "      <td>ENST00000294353</td>\n",
       "      <td>4.461</td>\n",
       "      <td>5.013</td>\n",
       "      <td>5.047</td>\n",
       "      <td>4.566</td>\n",
       "      <td>5.184</td>\n",
       "      <td>4.826</td>\n",
       "      <td>5.102</td>\n",
       "      <td>4.670</td>\n",
       "      <td>5.756</td>\n",
       "      <td>...</td>\n",
       "      <td>4.250</td>\n",
       "      <td>4.439</td>\n",
       "      <td>4.1460</td>\n",
       "      <td>0</td>\n",
       "      <td>201</td>\n",
       "      <td>201</td>\n",
       "      <td>2436</td>\n",
       "      <td>2436</td>\n",
       "      <td>8143</td>\n",
       "      <td>GGAGTCTGCGCTCTGGTTCGGGCTGCGGCTGCGGCTGCGGCTGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11562</th>\n",
       "      <td>ENST00000322764</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.524</td>\n",
       "      <td>5.478</td>\n",
       "      <td>5.915</td>\n",
       "      <td>5.811</td>\n",
       "      <td>5.817</td>\n",
       "      <td>5.943</td>\n",
       "      <td>5.509</td>\n",
       "      <td>4.931</td>\n",
       "      <td>...</td>\n",
       "      <td>5.598</td>\n",
       "      <td>5.968</td>\n",
       "      <td>5.3358</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>1799</td>\n",
       "      <td>1799</td>\n",
       "      <td>2228</td>\n",
       "      <td>GCAGAGTCTGCGGACCCGGCGCCGAGGCGGCCACCCGAGACGCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11563</th>\n",
       "      <td>ENST00000381638</td>\n",
       "      <td>5.112</td>\n",
       "      <td>4.918</td>\n",
       "      <td>5.139</td>\n",
       "      <td>5.190</td>\n",
       "      <td>5.442</td>\n",
       "      <td>5.602</td>\n",
       "      <td>4.715</td>\n",
       "      <td>4.956</td>\n",
       "      <td>5.033</td>\n",
       "      <td>...</td>\n",
       "      <td>5.038</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.0619</td>\n",
       "      <td>0</td>\n",
       "      <td>135</td>\n",
       "      <td>135</td>\n",
       "      <td>9021</td>\n",
       "      <td>9021</td>\n",
       "      <td>11466</td>\n",
       "      <td>AGGAAGCCGGAAGCCGCAGGGGCCGCCGTCGTCTCCTCCGCGTCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11564</th>\n",
       "      <td>ENST00000370801</td>\n",
       "      <td>3.165</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.539</td>\n",
       "      <td>3.629</td>\n",
       "      <td>3.616</td>\n",
       "      <td>4.203</td>\n",
       "      <td>3.400</td>\n",
       "      <td>NA</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>3.008</td>\n",
       "      <td>3.6041</td>\n",
       "      <td>0</td>\n",
       "      <td>476</td>\n",
       "      <td>476</td>\n",
       "      <td>3188</td>\n",
       "      <td>3188</td>\n",
       "      <td>6412</td>\n",
       "      <td>GTTGGCAGAGCAGTTGTCCTGGATGGCGGAGCCTTGGGTTCCGGGG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11565 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      EnsemblTranscriptID Adrenal_PTR Appendices_PTR Brain_PTR Colon_PTR  \\\n",
       "0         ENST00000263100          NA          8.277        NA        NA   \n",
       "1         ENST00000373993          NA             NA        NA     5.135   \n",
       "2         ENST00000318602       6.290          6.328     5.948     5.811   \n",
       "3         ENST00000299698          NA             NA     3.995        NA   \n",
       "4         ENST00000401850       3.843          4.601        NA        NA   \n",
       "...                   ...         ...            ...       ...       ...   \n",
       "11560     ENST00000374888          NA             NA        NA        NA   \n",
       "11561     ENST00000294353       4.461          5.013     5.047     4.566   \n",
       "11562     ENST00000322764       5.664          5.524     5.478     5.915   \n",
       "11563     ENST00000381638       5.112          4.918     5.139     5.190   \n",
       "11564     ENST00000370801       3.165             NA        NA     3.539   \n",
       "\n",
       "      Duodenum_PTR Endometrium_PTR Esophagus_PTR Fallopiantube_PTR Fat_PTR  \\\n",
       "0               NA              NA            NA             7.841      NA   \n",
       "1            5.371              NA            NA                NA      NA   \n",
       "2            6.068           5.383         5.881             6.119   6.410   \n",
       "3               NA              NA         4.129                NA      NA   \n",
       "4               NA              NA         4.013             3.683      NA   \n",
       "...            ...             ...           ...               ...     ...   \n",
       "11560           NA              NA            NA                NA      NA   \n",
       "11561        5.184           4.826         5.102             4.670   5.756   \n",
       "11562        5.811           5.817         5.943             5.509   4.931   \n",
       "11563        5.442           5.602         4.715             4.956   5.033   \n",
       "11564        3.629           3.616         4.203             3.400      NA   \n",
       "\n",
       "       ... Thyroid_PTR Tonsil_PTR Urinarybladder_PTR 5UTR start 5UTR stop  \\\n",
       "0      ...          NA         NA                 NA          0        55   \n",
       "1      ...          NA         NA                 NA          0        92   \n",
       "2      ...       6.060      5.675             5.8286          0        70   \n",
       "3      ...          NA      5.249                 NA          0        31   \n",
       "4      ...          NA         NA             4.2430          0       490   \n",
       "...    ...         ...        ...                ...        ...       ...   \n",
       "11560  ...          NA      4.681                 NA          0        38   \n",
       "11561  ...       4.250      4.439             4.1460          0       201   \n",
       "11562  ...       5.598      5.968             5.3358          0        80   \n",
       "11563  ...       5.038      5.130             5.0619          0       135   \n",
       "11564  ...          NA      3.008             3.6041          0       476   \n",
       "\n",
       "      CDS start CDS stop 3UTR start 3UTR stop  \\\n",
       "0            55     1543       1543      3382   \n",
       "1            92     1877       1877      2044   \n",
       "2            70     4495       4495      4610   \n",
       "3            31     4396       4396      5127   \n",
       "4           490     1552       1552      2321   \n",
       "...         ...      ...        ...       ...   \n",
       "11560        38     2450       2450      5467   \n",
       "11561       201     2436       2436      8143   \n",
       "11562        80     1799       1799      2228   \n",
       "11563       135     9021       9021     11466   \n",
       "11564       476     3188       3188      6412   \n",
       "\n",
       "                                              transcript  \n",
       "0      ATTGCTGCAGACGCTCACCCCAGACACTCACTGCACCGGAGTGAGC...  \n",
       "1      ATAATCAAGGAAACCTTTTCCGGGTGGGGATCTCTGAAATTACTCA...  \n",
       "2      GGGACCAGATGGATTGTAGGGAGTAGGGTACAATACAGTCTGTTCT...  \n",
       "3      GACCCTGGAAAAATCTGTCTCACCCACAAAGATGTGGGCTCAGCTC...  \n",
       "4      TGCACTTCTGTGCCTCAATTTCCTCATCTGTAGGGTGGGGGTGGTG...  \n",
       "...                                                  ...  \n",
       "11560  CTCGGCTCTGGTTCCAGCCGAGCCTCTCGGACGCAGAGATGGAAAT...  \n",
       "11561  GGAGTCTGCGCTCTGGTTCGGGCTGCGGCTGCGGCTGCGGCTGCGG...  \n",
       "11562  GCAGAGTCTGCGGACCCGGCGCCGAGGCGGCCACCCGAGACGCGGC...  \n",
       "11563  AGGAAGCCGGAAGCCGCAGGGGCCGCCGTCGTCTCCTCCGCGTCCC...  \n",
       "11564  GTTGGCAGAGCAGTTGTCCTGGATGGCGGAGCCTTGGGTTCCGGGG...  \n",
       "\n",
       "[11565 rows x 37 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# raw data file and path\n",
    "datafile = '../data/preproc_stage2.csv'\n",
    "\n",
    "# sanity check if the file exists\n",
    "if not Path(datafile).is_file():\n",
    "    print('Data file not found!')\n",
    "\n",
    "# reading the data into a dataframe and looking at the first entries\n",
    "df = pd.read_csv(datafile)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "03e7ccbb-28c7-4bb6-b86e-d659c45d0aae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS start 55 CDS stop 1543\n",
      "ATTGCTGCAGACGCTCACCCCAGACACTCACTGCACCGGAGTGAGCGCGACCATCATGTCCATGCTCGTGGTCTTTCTCTTGCTGTGGGGTGTCACCTGGGGCCCAGTGACAGAAGCAGCCATATTTTATGAGACGCAGCCCAGCCTGTGGGCAGAGTCCGAATCACTGCTGAAACCCTTGGCCAATGTGACGCTGACGTGCCAGGCCCACCTGGAGACTCCAGACTTCCAGCTGTTCAAGAATGGGGTGGCCCAGGAGCCTGTGCACCTTGACTCACCTGCCATCAAGCACCAGTTCCTGCTGACGGGTGACACCCAGGGCCGCTACCGCTGCCGCTCGGGCTTGTCCACAGGATGGACCCAGCTGAGCAAGCTCCTGGAGCTGACAGGGCCAAAGTCCTTGCCTGCTCCCTGGCTCTCGATGGCGCCAGTGTCCTGGATCACCCCCGGCCTGAAAACAACAGCAGTGTGCCGAGGTGTGCTGCGGGGTGTGACTTTTCTGCTGAGGCGGGAGGGCGACCATGAGTTTCTGGAGGTGCCTGAGGCCCAGGAGGATGTGGAGGCCACCTTTCCAGTCCATCAGCCTGGCAACTACAGCTGCAGCTACCGGACCGATGGGGAAGGCGCCCTCTCTGAGCCCAGCGCTACTGTGACCATTGAGGAGCTCGCTGCACCACCACCGCCTGTGCTGATGCACCATGGAGAGTCCTCCCAGGTCCTGCACCCTGGCAACAAGGTGACCCTCACCTGCGTGGCTCCCCTGAGTGGAGTGGACTTCCAGCTACGGCGCGGGGAGAAAGAGCTGCTGGTACCCAGGAGCAGCACCAGCCCAGATCGCATCTTCTTTCACCTGAACGCGGTGGCCCTGGGGGATGGAGGTCACTACACCTGCCGCTACCGGCTGCATGACAACCAAAACGGCTGGTCCGGGGACAGCGCGCCGGTCGAGCTGATTCTGAGCGATGAGACGCTGCCCGCGCCGGAGTTCTCCCCGGAGCCGGAGTCCGGCAGGGCCTTGCGGCTGCGGTGCCTGGCGCCCCTGGAGGGCGCGCGCTTCGCCCTGGTGCGCGAGGACAGGGGCGGGCGCCGCGTGCACCGTTTCCAGAGCCCCGCTGGGACCGAGGCGCTCTTCGAGCTGCACAACATTTCCGTGGCTGACTCCGCCAACTACAGCTGCGTCTACGTGGACCTGAAGCCGCCTTTCGGGGGCTCCGCGCCCAGCGAGCGCTTGGAGCTGCACGTGGACGGACCCCCTCCCAGGCCTCAGCTCCGGGCGACGTGGAGTGGGGCGGTCCTGGCGGGCCGAGATGCCGTCCTGCGCTGCGAGGGACCCATCCCCGACGTCACCTTCGAGCTGCTGCGCGAGGGCGAGACGAAGGCCGTGAAGACGGTCCGCACCCCCGGGGCCGCGGCGAACCTCGAGCTGATCTTCGTGGGGCCCCAGCACGCCGGCAACTACAGGTGCCGCTACCGCTCCTGGGTGCCCCACACCTTCGAATCGGAGCTCAGCGACCCTGTGGAGCTCCTGGTGGCAGAAAGCTGATGCAGCCGCGGGCCCAGGGTGCTGTTGGTGTCCTCAGAAGTGCCGGGGATTCTGGACTGGCTCCCTCCCCTCCTGTTGCAGCACAAGGCCGGGGTCTCTGGGGGGCTGGAGAAGCCTCCCTCATTCCTCCCAGGAATTAATAAATGTGAAGAGAGGCTCTGTTTAAAATGTCTTTGGACTCCCAGGGCTGAGTGGGCTGGGATCTCGTGTCCTCAAAGTGGATGGGTTCTGGGGTGGCTCCTGAGGTAGAGGAGTGGAGAACTGGCTCTTAAGAGCACAGTTTTCTTTTCTTTTTTCTTTTTTGAGACAGGGTCTCACACTGTCACCCAGGCTGGAGTGCAGTGTTGCAGTCCGGCTCACTGCAGTTTTGACCTCCCAGGCTCAAGCGATCCTCCTGCCTCAGCCTCCCAAGTAGCTGGGAGCCTGGGCATGCATCGCCACGTCTGGCTAATTATTATTTTTTGTAGACAGGGTCTCACTGTGTTGCCCAAGCTTGTCTTGAACTCCTGGCCTTAAGTGATCCTCCCACCTCAGCCTCCTGAGTAGCTGGGACTACAGGCATGAGCCACCATGCCTGGCCAACTCACATTTTTCTTTCTATTATTTATTTTTTGTAGAGATGAGTCTCACTATGTTGCCCAGGCTGGTCTTGAACTCTTGGGCTCAAGTGATCCTCCCGCCTCAGCCTCCCAAAGTGTTGGGATTACAGGTGTGAGCCACAGCACCTGGCCAACCAACACTTCTCAGGGCCTCTTTCATCTGTGCTCTTCCAGGATGCTGCCTCTTACTCCCTGGGCACCTCGGCCTGGTCCCAGCAGGTATGGGCAGTTGCTTGAGGCTCCAGACATACTCACCTCTACCTCGACCACATCAACCCCATCACCAAGAGGAGGTTCAGGGAAGCTGCATTTTGTGGTCTTGTCCTCCCAGTCCAGGGTGGTAGTGCTGGGCTGTTCCCAGCCTCCCACAGCAGAGCTGGACGCTGTGGAAATGGCTGGATTCCTCTGTGTTCTTTCCCATTATCCCTCAGCTCTGAGTCCTCTTGTGCCATCTGACATCTGATGCCTTCCCAACCACTGCCGAGTTTCTGCCTGGAGCAGGGCCTCAAGGCCCTGGCACACAGAAGATGCGTATCAGTATTATCAACCAATAGTTGATGAATTGTGTTTTTCAACGAATTTGCTAGTGATCTGGTTTACTGCCTTAGTAATATCTAGTTCCTAATATGCCTATGCCTTTTAATGTCTGCACAGTCTATGATGATGTCACTTCTCTCAAGCCTGATGTGTCCTCTCTCTCTTTCTTGCTAGTGGTTTATCAATTTTTAAAACCTTTTCTTGTTTATTTGTTTTTGAGACAGAATTTTGCTCTTGTTGCCCAGGCTGGAGTGCAATGGCGTGATCTTGGCTCACAGCAACCTCTACCTCCTGGGTTCAAGTGATTCTCGTGCCTCAGCCTCCTGAGTAGCTGGGCTGACAGGCACCCACCACCACACCCAGCTAATTTTTATGTTTTTAGTAGAGATGGGGTTTCACCTTGTTGGGTGGCCAGGCTGGTCTCCAACTCCTGAGCTCAAGTGATCCACCCGCCTCAGCCTCCCAAAGTGCTGGGATTACAGGGGTGAGCCACCACGCCCAGCCCCAGCTTAGTTTTTTAAAAAGTTTATTTTAGCCATTCTAATAGGTATGTAGACATATCTAATAGTGGTTTTCCTTGCGTTTCCTTAATGTCTGATGATGTTAAGCATTTTTTCCCCAAGTGCTTAGTTGCCATCTATATAGCATCTTTGATGAAATGTCTGTTTATATATTTTGCACACTTTAAAATATTGGGTTGTTT\n",
      "\n",
      "CDS structure ok.\n",
      "Start codon ['A', 'T', 'G'] Stop codon ['T', 'G', 'A']\n",
      "Start codon ok.\n",
      "Stop codon ok.\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "\n",
    "start_codons = [\n",
    "    ['A', 'T', 'G'],\n",
    "    ['C', 'T', 'G']\n",
    "]\n",
    "\n",
    "stop_codons = [\n",
    "    ['T', 'A', 'A'],\n",
    "    ['T', 'A', 'G'],\n",
    "    ['T', 'G', 'A']\n",
    "]\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    print('CDS start', row['CDS start'], 'CDS stop', row['CDS stop'])\n",
    "    print(row['transcript'])\n",
    "\n",
    "    CDS_length = row['CDS stop'] - row['CDS start']\n",
    "    if (CDS_length % 3) != 0:\n",
    "        print(row['EnsemblTranscriptID'], 'CDS length is NOT a multiple of three!')\n",
    "    else:\n",
    "        print('CDS structure ok.')\n",
    "\n",
    "    CDS = list(row['transcript'])\n",
    "    \n",
    "    CDS_start_codon = CDS[row['CDS start']:row['CDS start']+3]\n",
    "    CDS_stop_codon = CDS[row['CDS stop']-3:row['CDS stop']]\n",
    "\n",
    "    print('Start codon', CDS_start_codon, 'Stop codon', CDS_stop_codon)\n",
    "\n",
    "    if not CDS_start_codon in start_codons:\n",
    "        print(row['EnsemblTranscriptID'], 'Start codon invalid!', CDS_start_codon)\n",
    "    else:\n",
    "        print('Start codon ok.')\n",
    "\n",
    "    if not CDS_stop_codon in stop_codons:\n",
    "        print(row['EnsemblTranscriptID'], 'Stop codon invalid!', CDS_stop_codon)\n",
    "    else:\n",
    "        print('Stop codon ok.')\n",
    "    \n",
    "    count += 1\n",
    "    if count == 1: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aca8f374-c5a3-499d-858a-4bf58e8fbb08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 3, 4]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo = [1, 2, 3, 4, 5, 6]\n",
    "foo[4-3:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3cd29727-5ba1-4ff2-9e3f-b021554f56c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-11-19 23:32:53.887258: I tensorflow/core/util/port.cc:113] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.\n",
      "2023-11-19 23:32:53.888577: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2023-11-19 23:32:53.906128: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered\n",
      "2023-11-19 23:32:53.906145: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered\n",
      "2023-11-19 23:32:53.906631: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2023-11-19 23:32:53.909568: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2023-11-19 23:32:53.909848: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-11-19 23:32:54.478408: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[[1, 0, 0, 0],\n",
       "        [0, 1, 0, 0],\n",
       "        [0, 0, 1, 0],\n",
       "        [0, 0, 0, 1],\n",
       "        [0, 0, 0, 0],\n",
       "        [0, 0, 0, 0]],\n",
       "\n",
       "       [[1, 0, 0, 0],\n",
       "        [1, 0, 0, 0],\n",
       "        [1, 0, 0, 0],\n",
       "        [0, 0, 0, 0],\n",
       "        [0, 0, 0, 0],\n",
       "        [0, 0, 0, 0]],\n",
       "\n",
       "       [[0, 0, 1, 0],\n",
       "        [0, 0, 0, 1],\n",
       "        [0, 1, 0, 0],\n",
       "        [1, 0, 0, 0],\n",
       "        [1, 0, 0, 0],\n",
       "        [0, 1, 0, 0]],\n",
       "\n",
       "       [[0, 0, 0, 1],\n",
       "        [0, 0, 1, 0],\n",
       "        [1, 0, 0, 0],\n",
       "        [0, 1, 0, 0],\n",
       "        [0, 0, 0, 0],\n",
       "        [0, 0, 0, 0]]], dtype=int32)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from keras.utils import pad_sequences\n",
    "\n",
    "def seq_onehot(seq):\n",
    "    mapping = {\n",
    "        'A' : 0,\n",
    "        'T' : 1,\n",
    "        'G' : 2,\n",
    "        'C' : 3\n",
    "    }\n",
    "    seq2 = [ mapping[i] for i in seq ]\n",
    "    return np.eye(4)[seq2]\n",
    "\n",
    "raw_sequences = ['ATGC', 'AAA', 'GCTAAT', 'CGAT']\n",
    "\n",
    "ohe_sequences = [ seq_onehot(s) for s in raw_sequences ]\n",
    "pad_sequences(ohe_sequences, padding='post')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a8233e52-7b8f-4007-af5b-6a8109bd394d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "[[1 0 0 0]\n",
      " [0 1 0 0]\n",
      " [0 0 1 0]\n",
      " [0 0 0 1]]\n",
      "[[1 0 0 0]\n",
      " [0 1 0 0]\n",
      " [0 0 1 0]\n",
      " [0 0 0 1]\n",
      " [0 0 0 0]\n",
      " [0 0 0 0]]\n"
     ]
    }
   ],
   "source": [
    "foo = np.array(\n",
    "        [[1, 0, 0, 0],\n",
    "         [0, 1, 0, 0],\n",
    "         [0, 0, 1, 0],\n",
    "         [0, 0, 0, 1]],\n",
    "        dtype=int)\n",
    "\n",
    "print(len(foo))\n",
    "\n",
    "target = 6\n",
    "\n",
    "print(foo)\n",
    "foo = np.pad(foo, ((0,target-len(foo)), (0,0)), mode='constant')\n",
    "print(foo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "16c71ebb-6a9e-47c7-a70a-fbabeb459beb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def onehot_tissue(idx):\n",
    "    t = np.eye(29, dtype=int)[idx]\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "80733bf3-74ca-4e10-b216-e3a89f65799c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0]\n",
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]\n"
     ]
    }
   ],
   "source": [
    "for i in range(1, 30):\n",
    "    foo = i - 1\n",
    "    print(onehot_tissue(foo))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a86add1f-8bbe-4ae9-ba3e-843aeae38cd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hotenc(seq, u5_start=-1, u5_stop=-1, c_start=-1, c_stop=-1, u3_start=-1, u3_stop=-1, multi=True):\n",
    "    if multi: encoded_sequence = np.empty((len(seq),10), dtype=int)\n",
    "    else: encoded_sequence = np.empty((len(seq),4), dtype=int)\n",
    "    \n",
    "    for idx, x in enumerate(seq):\n",
    "        if multi: encoded = np.zeros((10,), dtype=int)\n",
    "        else: encoded = np.zeros((4,), dtype=int)\n",
    "        \n",
    "        # one hot encoding the sequence\n",
    "        mapping = { 'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3 }\n",
    "        encoded[0:4] = np.eye(4)[mapping[x]]\n",
    "\n",
    "        if multi:\n",
    "            # one hot encoding the region\n",
    "            if c_start != -1 and idx >= c_start and idx < c_stop: encoded[4:7] = [0, 1, 0]\n",
    "            elif u5_start != -1 and idx >= u5_start and idx < u5_stop: encoded[4:7] = [1, 0, 0]\n",
    "            elif u3_start != -1 and idx >= u3_start and idx < u3_stop: encoded[4:7] = [0, 0, 1]\n",
    "            else: encoded[4:7] = [0, 0, 0]\n",
    "\n",
    "            # one hot encoding the codon position\n",
    "            if idx >= c_start and idx < c_stop:\n",
    "                if u5_start == -1: u5_offset = 0\n",
    "                else: u5_offset = u5_stop\n",
    "                # get the position within the coding region by subtracting the 5'UTR offset\n",
    "                pos = (idx - u5_offset) % 3\n",
    "                encoded[7:10] = np.eye(3)[pos]\n",
    "            else:\n",
    "                encoded[7:10] = [0, 0, 0]      \n",
    "\n",
    "        encoded_sequence[idx] = encoded\n",
    "\n",
    "    return encoded_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71aecf11-68f4-4272-b98f-c39d0c430bb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 0, 0, 0, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 1, 0, 0, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 1, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 1, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 1, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 1, 0, 1, 0, 0, 0, 0, 0],\n",
       "       [1, 0, 0, 0, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 1, 0, 0, 1, 0, 1, 0, 0],\n",
       "       [0, 1, 0, 0, 0, 1, 0, 0, 1, 0],\n",
       "       [0, 1, 0, 0, 0, 1, 0, 0, 0, 1],\n",
       "       [0, 1, 0, 0, 0, 1, 0, 1, 0, 0],\n",
       "       [0, 0, 0, 1, 0, 1, 0, 0, 1, 0],\n",
       "       [0, 0, 1, 0, 0, 1, 0, 0, 0, 1],\n",
       "       [1, 0, 0, 0, 0, 1, 0, 1, 0, 0],\n",
       "       [0, 1, 0, 0, 0, 1, 0, 0, 1, 0],\n",
       "       [1, 0, 0, 0, 0, 1, 0, 0, 0, 1],\n",
       "       [0, 0, 1, 0, 0, 0, 1, 0, 0, 0],\n",
       "       [0, 0, 1, 0, 0, 0, 1, 0, 0, 0],\n",
       "       [0, 1, 0, 0, 0, 0, 1, 0, 0, 0],\n",
       "       [0, 0, 1, 0, 0, 0, 1, 0, 0, 0]])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "UTR5_start = 0\n",
    "UTR5_stop = 7\n",
    "CDS_start = 7\n",
    "CDS_stop = 16\n",
    "UTR3_start = 16\n",
    "UTR3_stop = 21\n",
    "\n",
    "raw_sequence = 'ATCCCGAGTTTCGATAGGTG'\n",
    "\n",
    "hotenc(raw_sequence, UTR5_start, UTR5_stop, CDS_start, CDS_stop, UTR3_start, UTR3_stop, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c0025fd9-30be-4dee-ace7-59df13c9140e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'I': 3, 'am': 2, 'fourty': 1, 'five': 1, 'years': 1, 'old.': 1, 'People': 1, 'suck.': 1, 'do': 1, 'not': 1, 'give': 1, 'a': 1, 'fuck.': 1, 'with': 1, 'my': 2, 'dogs': 1, 'who': 1, 'are': 1, 'best': 1, 'friends.': 1}\n",
      "20\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[14,\n",
       " 8,\n",
       " 3,\n",
       " 9,\n",
       " 3,\n",
       " 12,\n",
       " 15,\n",
       " 3,\n",
       " 14,\n",
       " 1,\n",
       " 11,\n",
       " 9,\n",
       " 16,\n",
       " 15,\n",
       " 14,\n",
       " 8,\n",
       " 5,\n",
       " 5,\n",
       " 1,\n",
       " 17,\n",
       " 8,\n",
       " 5,\n",
       " 11,\n",
       " 14]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from keras.preprocessing.text import one_hot\n",
    "\n",
    "text = \"I am fourty five years old. People suck. I do not give a fuck. I am with my dogs who are my best friends.\"\n",
    "\n",
    "wc = {}\n",
    "for w in text.split():\n",
    "    # print(w)\n",
    "    if w in wc.keys():\n",
    "        wc[w] += 1\n",
    "    else:\n",
    "        wc[w] = 1\n",
    "\n",
    "voc_size = len(wc.keys())\n",
    "print(wc)\n",
    "print(voc_size)\n",
    "\n",
    "one_hot(text, voc_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5aaff88e-6f62-4ebb-a361-8de8c8770abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_kmers(seq, ksize):\n",
    "    voc = {}\n",
    "    kmer_count = (len(seq) - ksize) + 1\n",
    "    print('k-mer count:', kmer_count)\n",
    "    \n",
    "    for i in range(0, kmer_count):\n",
    "        kmer = seq[i:i+ksize]\n",
    "        if kmer in voc.keys():\n",
    "            voc[kmer] += 1\n",
    "        else:\n",
    "            voc[kmer] = 1\n",
    "\n",
    "    return voc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "824654f9-09d7-4125-88c6-29152734bad6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence length: 4203\n",
      "k-mer count: 4201\n",
      "vocabulary length: 64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'AGA': 86,\n",
       " 'GAA': 57,\n",
       " 'AAG': 80,\n",
       " 'AGG': 119,\n",
       " 'GGC': 148,\n",
       " 'GCA': 83,\n",
       " 'CAG': 109,\n",
       " 'AGC': 87,\n",
       " 'GCC': 143,\n",
       " 'CCT': 127,\n",
       " 'CTC': 93,\n",
       " 'TCG': 41,\n",
       " 'CGG': 91,\n",
       " 'GGT': 51,\n",
       " 'GTC': 44,\n",
       " 'TCT': 43,\n",
       " 'CTG': 144,\n",
       " 'TGG': 99,\n",
       " 'GGG': 153,\n",
       " 'GCG': 67,\n",
       " 'CCC': 161,\n",
       " 'CCA': 99,\n",
       " 'CAC': 75,\n",
       " 'ACT': 34,\n",
       " 'TGC': 89,\n",
       " 'CCG': 92,\n",
       " 'CGC': 70,\n",
       " 'GCT': 101,\n",
       " 'TGT': 42,\n",
       " 'GTG': 81,\n",
       " 'TGA': 56,\n",
       " 'GAC': 104,\n",
       " 'ACC': 103,\n",
       " 'ACA': 48,\n",
       " 'CAT': 48,\n",
       " 'ATG': 34,\n",
       " 'GAG': 119,\n",
       " 'CGT': 44,\n",
       " 'GGA': 110,\n",
       " 'GAT': 28,\n",
       " 'CGA': 56,\n",
       " 'CTT': 40,\n",
       " 'TTC': 41,\n",
       " 'TCC': 72,\n",
       " 'GTT': 24,\n",
       " 'TTT': 40,\n",
       " 'TTG': 27,\n",
       " 'CTA': 28,\n",
       " 'TAC': 30,\n",
       " 'ACG': 61,\n",
       " 'CAA': 64,\n",
       " 'AAA': 40,\n",
       " 'AGT': 27,\n",
       " 'TCA': 66,\n",
       " 'ATC': 45,\n",
       " 'AAC': 37,\n",
       " 'GTA': 15,\n",
       " 'TAT': 9,\n",
       " 'AAT': 18,\n",
       " 'ATA': 10,\n",
       " 'TAA': 14,\n",
       " 'ATT': 14,\n",
       " 'TAG': 10,\n",
       " 'TTA': 10}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_sequence = 'ATGCTGA'\n",
    "raw_sequence = 'ATCCCGAGTTTCGATAGGTG'\n",
    "raw_sequence = 'AGAAGGCAGCCTCGGTCTCTGGGCGGCGGCGGCGGCCCACTCTGCCCTGGCCGCGCTGTGTGGTGACCGCAGGCCCCAGACATGAGGGCGGCCCGTGCTCTGCTGCCCCTGCTGCTGCAGGCCTGCTGGACAGCCGCGCAGGATGAGCCGGAGACCCCGAGGGCCGTGGCCTTCCAGGACTGCCCCGTGGACCTGTTCTTTGTGCTGGACACCTCTGAGAGCGTGGCCCTGAGGCTGAAGCCCTACGGGGCCCTCGTGGACAAAGTCAAGTCCTTCACCAAGCGCTTCATCGACAACCTGAGGGACAGGTACTACCGCTGTGACCGAAACCTGGTGTGGAACGCAGGCGCGCTGCACTACAGTGACGAGGTGGAGATCATCCAAGGCCTCACGCGCATGCCTGGCGGCCGCGACGCACTCAAAAGCAGCGTGGACGCGGTCAAGTACTTTGGGAAGGGCACCTACACCGACTGCGCTATCAAGAAGGGGCTGGAGCAGCTCCTCGTGGGGGGCTCCCACCTGAAGGAGAATAAGTACCTGATTGTGGTGACCGACGGGCACCCCCTGGAGGGCTACAAGGAACCCTGTGGGGGGCTGGAGGATGCTGTGAACGAGGCCAAGCACCTGGGCGTCAAAGTCTTCTCGGTGGCCATCACACCCGACCACCTGGAGCCGCGTCTGAGCATCATCGCCACGGACCACACGTACCGGCGCAACTTCACGGCGGCTGACTGGGGCCAGAGCCGCGACGCAGAGGAGGCCATCAGCCAGACCATCGACACCATCGTGGACATGATCAAAAATAACGTGGAGCAAGTGTGCTGCTCCTTCGAATGCCAGCCTGCAAGAGGACCTCCGGGGCTCCGGGGCGACCCCGGCTTTGAGGGAGAACGAGGCAAGCCGGGGCTCCCAGGAGAGAAGGGAGAAGCCGGAGATCCTGGAAGACCCGGGGACCTCGGACCTGTTGGGTACCAGGGAATGAAGGGAGAAAAAGGGAGCCGTGGGGAGAAGGGCTCCAGGGGACCCAAGGGCTACAAGGGAGAGAAGGGCAAGCGTGGCATCGACGGGGTGGACGGCGTGAAGGGGGAGATGGGGTACCCAGGCCTGCCAGGCTGCAAGGGCTCGCCCGGGTTTGACGGCATTCAAGGACCCCCTGGCCCCAAGGGAGACCCCGGTGCCTTTGGACTGAAAGGAGAAAAGGGCGAGCCTGGAGCTGACGGGGAGGCGGGGAGACCAGGGAGCTCGGGACCATCTGGAGACGAGGGCCAGCCGGGAGAGCCTGGGCCCCCCGGAGAGAAAGGAGAGGCGGGCGACGAGGGGAACCCAGGACCTGACGGTGCCCCCGGGGAGCGGGGTGGCCCTGGAGAGAGAGGACCACGGGGGACCCCAGGCACGCGGGGACCAAGAGGAGACCCTGGTGAAGCTGGCCCGCAGGGTGATCAGGGAAGAGAAGGCCCCGTTGGTGTCCCTGGAGACCCGGGCGAGGCTGGCCCTATCGGACCTAAAGGCTACCGAGGCGATGAGGGTCCCCCAGGGTCCGAGGGTGCCAGAGGAGCCCCAGGACCTGCCGGACCCCCTGGAGACCCGGGGCTGATGGGTGAAAGGGGAGAAGACGGCCCCGCTGGAAATGGCACCGAGGGCTTCCCCGGCTTCCCCGGGTATCCGGGCAACAGGGGCGCTCCCGGGATAAACGGCACGAAGGGCTACCCCGGCCTCAAGGGGGACGAGGGAGAAGCCGGGGACCCCGGAGACGATAACAACGACATTGCACCCCGAGGAGTCAAAGGAGCAAAGGGGTACCGGGGTCCCGAGGGCCCCCAGGGACCCCCAGGACACCAAGGACCGCCTGGGCCGGACGAATGCGAGATTTTGGACATCATCATGAAAATGTGCTCTTGCTGTGAATGCAAGTGCGGCCCCATCGACCTCCTGTTCGTGCTGGACAGCTCAGAGAGCATTGGCCTGCAGAACTTCGAGATTGCCAAGGACTTCGTCGTCAAGGTCATCGACCGGCTGAGCCGGGACGAGCTGGTCAAGTTCGAGCCAGGGCAGTCGTACGCGGGTGTGGTGCAGTACAGCCACAGCCAGATGCAGGAGCACGTGAGCCTGCGCAGCCCCAGCATCCGGAACGTGCAGGAGCTCAAGGAAGCCATCAAGAGCCTGCAGTGGATGGCGGGCGGCACCTTCACGGGGGAGGCCCTGCAGTACACGCGGGACCAGCTGCTGCCGCCCAGCCCGAACAACCGCATCGCCCTGGTCATCACTGACGGGCGCTCAGACACTCAGAGGGACACCACACCGCTCAACGTGCTCTGCAGCCCCGGCATCCAGGTGGTCTCCGTGGGCATCAAAGACGTGTTTGACTTCATCCCAGGCTCAGACCAGCTCAATGTCATTTCTTGCCAAGGCCTGGCACCATCCCAGGGCCGGCCCGGCCTCTCGCTGGTCAAGGAGAACTATGCAGAGCTGCTGGAGGATGCCTTCCTGAAGAATGTCACCGCCCAGATCTGCATAGACAAGAAGTGTCCAGATTACACCTGCCCCATCACGTTCTCCTCCCCGGCTGACATCACCATCCTGCTGGACGGCTCCGCCAGCGTGGGCAGCCACAACTTTGACACCACCAAGCGCTTCGCCAAGCGCCTGGCCGAGCGCTTCCTCACAGCGGGCAGGACGGACCCCGCCCACGACGTGCGGGTGGCGGTGGTGCAGTACAGCGGCACGGGCCAGCAGCGCCCAGAGCGGGCGTCGCTGCAGTTCCTGCAGAACTACACGGCCCTGGCCAGTGCCGTCGATGCCATGGACTTTATCAACGACGCCACCGACGTCAACGATGCCCTGGGCTATGTGACCCGCTTCTACCGCGAGGCCTCGTCCGGCGCTGCCAAGAAGAGGCTGCTGCTCTTCTCAGATGGCAACTCGCAGGGCGCCACGCCCGCTGCCATCGAGAAGGCCGTGCAGGAAGCCCAGCGGGCAGGCATCGAGATCTTCGTGGTGGTCGTGGGCCGCCAGGTGAATGAGCCCCACATCCGCGTCCTGGTCACCGGCAAGACGGCCGAGTACGACGTGGCCTACGGCGAGAGCCACCTGTTCCGTGTCCCCAGCTACCAGGCCCTGCTCCGCGGTGTCTTCCACCAGACAGTCTCCAGGAAGGTGGCGCTGGGCTAGCCCACCCTGCACGCCGGCACCAAACCCTGTCCTCCCACCCCTCCCCACTCATCACTAAACAGAGTAAAATGTGATGCGAATTTTCCCGACCAACCTGATTCGCTAGATTTTTTTTAAGGAAAAGCTTGGAAAGCCAGGACACAACGCTGCTGCCTGCTTTGTGCAGGGTCCTCCGGGGCTCAGCCCTGAGTTGGCATCACCTGCGCAGGGCCCTCTGGGGCTCAGCCCTGAGCTAGTGTCACCTGCACAGGGCCCTCTGAGGCTCAGCCCTGAGCTGGCGTCACCTGTGCAGGGCCCTCTGGGGCTCAGCCCTGAGCTGGCCTCACCTGGGTTCCCCACCCCGGGCTCTCCTGCCCTGCCCTCCTGCCCGCCCTCCCTCCTGCCTGCGCAGCTCCTTCCCTAGGCACCTCTGTGCTGCATCCCACCAGCCTGAGCAAGACGCCCTCTCGGGGCCTGTGCCGCACTAGCCTCCCTCTCCTCTGTCCCCATAGCTGGTTTTTCCCACCAATCCTCACCTAACAGTTACTTTACAATTAAACTCAAAGCAAGCTCTTCTCCTCAGCTTGGGGCAGCCATTGGCCTCTGTCTCGTTTTGGGAAACCAAGGTCAGGAGGCCGTTGCAGACATAAATCTCGGCGACTCGGCCCCGTCTCCTGAGGGTCCTGCTGGTGACCGGCCTGGACCTTGGCCCTACAGCCCTGGAGGCCGCTGCTGACCAGCACTGACCCCGACCTCAGAGAGTACTCGCAGGGGCGCTGGCTGCACTCAAGACCCTCGAGATTAACGGTGCTAACCCCGTCTGCTCCTCCCTCCCGCAGAGACTGGGGCCTGGACTGGACATGAGAGCCCCTTGGTGCCACAGAGGGCTGTGTCTTACTAGAAACAACGCAAACCTCTCCTTCCTCAGAATAGTGATGTGTTCGACGTTTTATCAAAGGCCCCCTTTCTATGTTCATGTTAGTTTTGCTCCTTCTGTGTTTTTTTCTGAACCATATCCATGTTGCTGACTTTTCCAAATAAAGGTTTTCACTCCTC'\n",
    "\n",
    "print('sequence length:', len(raw_sequence))\n",
    "\n",
    "v = build_kmers(raw_sequence, 3)\n",
    "print('vocabulary length:', len(v.keys()))\n",
    "v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80fea491-c02a-4b96-bb61-5cd9c9237d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['AAA',\n",
       " 'AAT',\n",
       " 'AAG',\n",
       " 'AAC',\n",
       " 'ATA',\n",
       " 'ATT',\n",
       " 'ATG',\n",
       " 'ATC',\n",
       " 'AGA',\n",
       " 'AGT',\n",
       " 'AGG',\n",
       " 'AGC',\n",
       " 'ACA',\n",
       " 'ACT',\n",
       " 'ACG',\n",
       " 'ACC',\n",
       " 'TAA',\n",
       " 'TAT',\n",
       " 'TAG',\n",
       " 'TAC',\n",
       " 'TTA',\n",
       " 'TTT',\n",
       " 'TTG',\n",
       " 'TTC',\n",
       " 'TGA',\n",
       " 'TGT',\n",
       " 'TGG',\n",
       " 'TGC',\n",
       " 'TCA',\n",
       " 'TCT',\n",
       " 'TCG',\n",
       " 'TCC',\n",
       " 'GAA',\n",
       " 'GAT',\n",
       " 'GAG',\n",
       " 'GAC',\n",
       " 'GTA',\n",
       " 'GTT',\n",
       " 'GTG',\n",
       " 'GTC',\n",
       " 'GGA',\n",
       " 'GGT',\n",
       " 'GGG',\n",
       " 'GGC',\n",
       " 'GCA',\n",
       " 'GCT',\n",
       " 'GCG',\n",
       " 'GCC',\n",
       " 'CAA',\n",
       " 'CAT',\n",
       " 'CAG',\n",
       " 'CAC',\n",
       " 'CTA',\n",
       " 'CTT',\n",
       " 'CTG',\n",
       " 'CTC',\n",
       " 'CGA',\n",
       " 'CGT',\n",
       " 'CGG',\n",
       " 'CGC',\n",
       " 'CCA',\n",
       " 'CCT',\n",
       " 'CCG',\n",
       " 'CCC']"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# build the embedding vocabulary\n",
    "\n",
    "import itertools\n",
    "\n",
    "perms = [''.join(x) for x in itertools.product('ATGC', repeat=3)]\n",
    "\n",
    "print(len(perms))\n",
    "perms"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
